// @hcengineering/card/lib/analytics.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var analytics_exports = {};
__export(analytics_exports, {
  CardEvents: () => CardEvents
});
module.exports = __toCommonJS(analytics_exports);
var CardEvents = /* @__PURE__ */ ((CardEvents2) => {
  CardEvents2["TypeCreated"] = "card.card.TypeCreated";
  CardEvents2["TagCreated"] = "card.card.TagCreated";
  CardEvents2["RelationCreated"] = "card.card.RelationCreated";
  CardEvents2["CardCreated"] = "card.card.Created";
  CardEvents2["CardOpened"] = "card.card.Opened";
  CardEvents2["TypeChanged"] = "card.card.TypeChanged";
  return CardEvents2;
})(CardEvents || {});
