// @hcengineering/card/lib/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __reExport = (target, mod, secondTarget) => (__copyProps(target, mod, "default"), secondTarget && __copyProps(secondTarget, mod, "default"));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var index_exports = {};
__export(index_exports, {
  DOMAIN_CARD: () => DOMAIN_CARD,
  cardId: () => cardId,
  default: () => index_default
});
module.exports = __toCommonJS(index_exports);
var import_platform = require("@hcengineering/platform");
__reExport(index_exports, require("./analytics"), module.exports);
const cardId = "card";
const DOMAIN_CARD = "card";
const cardPlugin = (0, import_platform.plugin)(cardId, {
  class: {
    Card: "",
    MasterTag: "",
    Tag: "",
    MasterTagEditorSection: "",
    CardSpace: "",
    Role: "",
    CardSection: "",
    FavoriteCard: ""
  },
  mixin: {
    CardViewDefaults: "",
    CreateCardExtension: ""
  },
  space: {
    Default: ""
  },
  types: {
    File: "",
    Document: ""
  },
  icon: {
    MasterTags: "",
    MasterTag: "",
    Tag: "",
    Tags: "",
    Card: "",
    File: "",
    View: "",
    Document: "",
    Home: "",
    Space: "",
    Expand: "",
    Feed: "",
    All: ""
  },
  extensions: {
    EditCardExtension: ""
  },
  string: {
    MasterTag: "",
    MasterTags: "",
    Tags: "",
    Tag: "",
    Card: "",
    Cards: "",
    CardApplication: "",
    Views: "",
    Labels: "",
    GetIndividualPublicLink: "",
    AddTag: "",
    Feed: "",
    AllCards: ""
  },
  section: {
    Attachments: "",
    Children: "",
    Content: "",
    Properties: "",
    Relations: ""
  },
  ids: {
    CardWidget: ""
  },
  component: {
    LabelsPresenter: "",
    CardTagColored: "",
    CardTagsColored: "",
    CardIcon: "",
    CardArrayEditor: "",
    CardFeedView: ""
  },
  function: {
    OpenCardInSidebar: "",
    GetSpaceAccessPublicLink: "",
    CanGetSpaceAccessPublicLink: ""
  },
  label: {
    Subscribed: "",
    NewMessages: ""
  }
});
var index_default = cardPlugin;
