// @hcengineering/workbench/lib/plugin.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var plugin_exports = {};
__export(plugin_exports, {
  workbenchId: () => workbenchId,
  workbenchPlugin: () => workbenchPlugin
});
module.exports = __toCommonJS(plugin_exports);
var import_platform = require("@hcengineering/platform");
const workbenchId = "workbench";
const workbenchPlugin = (0, import_platform.plugin)(workbenchId, {
  class: {
    Application: "",
    ApplicationNavModel: "",
    HiddenApplication: "",
    Widget: "",
    WidgetPreference: "",
    WorkbenchTab: ""
  },
  mixin: {
    SpaceView: ""
  },
  action: {
    CloseCurrentTab: ""
  },
  component: {
    WorkbenchApp: "",
    InviteLink: "",
    Archive: "",
    SpecialView: ""
  },
  string: {
    Archive: "",
    View: "",
    ServerUnderMaintenance: "",
    UpgradeDownloadProgress: "",
    OpenInSidebar: "",
    OpenInSidebarNewTab: "",
    ConfigureWidgets: "",
    WorkspaceIsArchived: "",
    WorkspaceIsMigrating: ""
  },
  icon: {
    Search: ""
  },
  event: {
    NotifyConnection: "",
    NotifyTitle: ""
  },
  metadata: {
    PlatformTitle: "",
    ExcludedApplications: "",
    ExcludedApplicationsForAnonymous: "",
    DefaultApplication: "",
    DefaultSpace: "",
    DefaultSpecial: "",
    // Default for navigation expanded state
    NavigationExpandedDefault: ""
  },
  extensions: {
    WorkbenchExtensions: "",
    WorkbenchTabExtensions: "",
    SpecialViewAction: ""
  },
  function: {
    CreateWidgetTab: "",
    CloseWidgetTab: "",
    CloseWidget: "",
    GetSidebarObject: "",
    LogIn: "",
    LogOut: ""
  },
  actionImpl: {
    Navigate: ""
  }
});
