// @hcengineering/api-client/lib/socket/node.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var node_exports = {};
__export(node_exports, {
  NodeWebSocketFactory: () => NodeWebSocketFactory
});
module.exports = __toCommonJS(node_exports);
const NodeWebSocketFactory = /* @__PURE__ */ __name((url) => {
  let WebSocket;
  try {
    WebSocket = require("ws");
  } catch (error) {
    throw new Error('The "ws" package is required for NodeWebSocketFactory. ');
  }
  const ws = new WebSocket(url);
  const client = {
    get readyState() {
      return ws.readyState;
    },
    send: /* @__PURE__ */ __name((data) => {
      if (data instanceof Blob) {
        void data.arrayBuffer().then((buffer) => {
          ws.send(buffer);
        });
      } else {
        ws.send(data);
      }
    }, "send"),
    close: /* @__PURE__ */ __name((code) => {
      ws.close(code);
    }, "close")
  };
  ws.on("message", (data) => {
    if (client.onmessage != null) {
      const event = {
        data,
        type: "message",
        target: void 0
      };
      client.onmessage(event);
    }
  });
  ws.on("close", (code, reason) => {
    if (client.onclose != null) {
      const closeEvent = {
        code,
        reason,
        wasClean: code === 1e3,
        type: "close",
        target: void 0
      };
      client.onclose(closeEvent);
    }
  });
  ws.on("open", () => {
    if (client.onopen != null) {
      const event = {
        type: "open",
        target: void 0
      };
      client.onopen(event);
    }
  });
  ws.on("error", (error) => {
    if (client.onerror != null) {
      const event = {
        type: "error",
        target: void 0,
        error
      };
      client.onerror(event);
    }
  });
  return client;
}, "NodeWebSocketFactory");
