// @hcengineering/contact/lib/__tests__/cache.test.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var import_core = __toESM(require("@hcengineering/core"));
var import__ = __toESM(require(".."));
var import_cache = __toESM(require("../cache"));
describe("ContactCache", () => {
  let cache;
  let personId;
  let personRef;
  let person;
  let socialId;
  beforeEach(() => {
    ;
    import_cache.default._instance = void 0;
    cache = import_cache.default.instance;
    personId = "personId1";
    personRef = "personRef1";
    socialId = {
      _id: personId,
      attachedTo: personRef,
      attachedToClass: import__.default.class.Person,
      _class: import__.default.class.SocialIdentity,
      space: import__.default.space.Contacts,
      modifiedOn: 0,
      createdBy: personId,
      modifiedBy: personId,
      type: import_core.SocialIdType.EMAIL,
      value: "tester@huly.me",
      key: "email:tester@huly.me",
      collection: "socialIds"
    };
    person = {
      _id: personRef,
      _class: import__.default.class.Person,
      space: import__.default.space.Contacts,
      modifiedOn: 0,
      createdBy: personId,
      modifiedBy: personId,
      name: "Test Person",
      avatarType: import__.AvatarType.COLOR
    };
  });
  describe("fillCachesForPersonId", () => {
    it("should handle null social identity", () => {
      cache.fillCachesForPersonId(personId, null);
      expect(cache.personRefByPersonId.get(personId)).toBeNull();
      expect(cache.personByPersonId.get(personId)).toBeNull();
      expect(cache.socialIdByPersonId.get(personId)).toBeNull();
    });
    it("should fill caches with social identity without lookup", () => {
      cache.fillCachesForPersonId(personId, socialId);
      expect(cache.personRefByPersonId.get(personId)).toBe(personRef);
      expect(cache.personByPersonId.get(personId)).toBeNull();
      expect(cache.socialIdByPersonId.get(personId)).toBe(socialId);
    });
    it("should fill caches with social identity with person lookup", () => {
      const socialIdWLookup = {
        ...socialId,
        $lookup: {
          attachedTo: person
        }
      };
      cache.fillCachesForPersonId(personId, socialIdWLookup);
      expect(cache.personRefByPersonId.get(personId)).toBe(personRef);
      expect(cache.personByPersonId.get(personId)).toBe(person);
      expect(cache.socialIdByPersonId.get(personId)).toBe(socialIdWLookup);
      expect(cache.personByRef.get(personRef)).toBe(person);
    });
  });
  describe("fillCachesForPersonRef", () => {
    const personId2 = "personId2";
    it("should handle null person", () => {
      cache.fillCachesForPersonRef(personRef, null);
      expect(cache.personByRef.get(personRef)).toBeNull();
    });
    it("should fill caches with person and social identities", () => {
      const personWLookup = {
        ...person,
        $lookup: {
          socialIds: [
            socialId,
            {
              _id: personId2,
              attachedTo: personRef,
              _class: import__.default.class.SocialIdentity,
              space: import__.default.space.Contacts,
              modifiedOn: 0,
              createdBy: personId,
              modifiedBy: personId,
              attachedToClass: import__.default.class.Person,
              type: import_core.SocialIdType.EMAIL,
              value: "tester2@huly.me",
              key: "email:tester2@huly.me",
              collection: "socialIds"
            }
          ]
        }
      };
      cache.fillCachesForPersonRef(personRef, personWLookup);
      expect(cache.personByRef.get(personRef)).toBe(personWLookup);
      expect(cache.personRefByPersonId.get(personId)).toBe(personRef);
      expect(cache.personRefByPersonId.get(personId2)).toBe(personRef);
      expect(cache.personByPersonId.get(personId)).toBe(personWLookup);
      expect(cache.personByPersonId.get(personId2)).toBe(personWLookup);
    });
  });
  describe("handleTx", () => {
    let changeListener;
    beforeEach(() => {
      changeListener = jest.fn();
      cache.addChangeListener(changeListener);
    });
    afterEach(() => {
      cache.removeChangeListener(changeListener);
    });
    it("should handle person creation", () => {
      const createTx = {
        _id: "tx1",
        _class: import_core.default.class.TxCreateDoc,
        space: import__.default.space.Contacts,
        modifiedOn: 0,
        createdBy: personId,
        modifiedBy: personId,
        objectId: personRef,
        objectClass: import__.default.class.Person,
        objectSpace: import__.default.space.Contacts,
        attributes: {
          name: "Tester",
          avatarType: import__.AvatarType.COLOR
        }
      };
      cache.fillCachesForPersonRef(personRef, null);
      cache.handleTx([createTx]);
      expect(changeListener).toHaveBeenCalledWith({
        personRef,
        personIds: []
      });
      expect(cache.personByRef.get(personRef)?.name).toBe("Tester");
    });
    it("should handle social identity creation", () => {
      const createTx = {
        _id: "tx2",
        _class: import_core.default.class.TxCreateDoc,
        space: import__.default.space.Contacts,
        modifiedOn: 0,
        createdBy: personId,
        modifiedBy: personId,
        objectId: personId,
        objectClass: import__.default.class.SocialIdentity,
        objectSpace: import__.default.space.Contacts,
        attachedTo: personRef,
        attributes: {
          attachedTo: personRef,
          attachedToClass: import__.default.class.Person,
          type: import_core.SocialIdType.EMAIL,
          value: "tester@huly.me",
          key: "email:tester@huly.me",
          collection: "socialIds"
        }
      };
      cache.fillCachesForPersonRef(personRef, {
        ...person,
        $lookup: {
          socialIds: [socialId]
        }
      });
      cache.handleTx([createTx]);
      expect(changeListener).toHaveBeenCalledWith({
        personRef,
        personIds: [personId]
      });
    });
    it("should handle person update", () => {
      const updateTx = {
        _id: "tx3",
        _class: import_core.default.class.TxUpdateDoc,
        space: import__.default.space.Contacts,
        modifiedOn: 0,
        modifiedBy: personId,
        objectId: personRef,
        objectClass: import__.default.class.Person,
        operations: {},
        objectSpace: import__.default.space.Contacts
      };
      cache.fillCachesForPersonRef(personRef, person);
      cache.handleTx([updateTx]);
      expect(changeListener).toHaveBeenCalledWith({
        personRef,
        personIds: []
      });
    });
  });
  describe("change listeners", () => {
    it("should add and remove listeners", () => {
      const listener1 = jest.fn();
      const listener2 = jest.fn();
      cache.addChangeListener(listener1);
      cache.addChangeListener(listener2);
      const change = {
        personRef,
        personIds: []
      };
      const createTx = {
        _id: "tx1",
        _class: import_core.default.class.TxCreateDoc,
        space: import__.default.space.Contacts,
        modifiedOn: 0,
        createdBy: personId,
        modifiedBy: personId,
        objectId: personRef,
        objectClass: import__.default.class.Person,
        objectSpace: import__.default.space.Contacts,
        attributes: {
          name: "Tester",
          avatarType: import__.AvatarType.COLOR
        }
      };
      cache.fillCachesForPersonRef(personRef, null);
      cache.handleTx([createTx]);
      expect(listener1).toHaveBeenCalledWith(expect.objectContaining(change));
      expect(listener2).toHaveBeenCalledWith(expect.objectContaining(change));
      cache.removeChangeListener(listener1);
      const updateTx = {
        _id: "tx3",
        _class: import_core.default.class.TxUpdateDoc,
        space: import__.default.space.Contacts,
        modifiedOn: 0,
        modifiedBy: personId,
        objectId: personRef,
        objectClass: import__.default.class.Person,
        operations: {},
        objectSpace: import__.default.space.Contacts
      };
      cache.handleTx([updateTx]);
      expect(listener1).toHaveBeenCalledTimes(1);
      expect(listener2).toHaveBeenCalledTimes(2);
    });
  });
});
