// @hcengineering/preference/lib/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var index_exports = {};
__export(index_exports, {
  DOMAIN_PREFERENCE: () => DOMAIN_PREFERENCE,
  default: () => index_default,
  preferenceId: () => preferenceId
});
module.exports = __toCommonJS(index_exports);
var import_platform = require("@hcengineering/platform");
const preferenceId = "preference";
const DOMAIN_PREFERENCE = "preference";
const preference = (0, import_platform.plugin)(preferenceId, {
  class: {
    Preference: "",
    SpacePreference: ""
  },
  icon: {
    Star: ""
  },
  string: {
    Starred: "",
    Star: "",
    Unstar: ""
  }
});
var index_default = preference;
