// @hcengineering/text-html/lib/parser.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var parser_exports = {};
__export(parser_exports, {
  HtmlParser: () => HtmlParser
});
module.exports = __toCommonJS(parser_exports);
var import_text_core = require("@hcengineering/text-core");
var import_htmlparser2 = require("htmlparser2");
class HtmlParseState {
  constructor(root, handlers) {
    this.root = root;
    this.handlers = handlers;
    this.stack.push(root);
  }
  static {
    __name(this, "HtmlParseState");
  }
  stack = [];
  marks = [];
  top() {
    return this.stack[this.stack.length - 1];
  }
  addText(text) {
    const top = this.top();
    if (top === void 0 || text.length === 0) {
      return;
    }
    if (top.type === import_text_core.MarkupNodeType.doc) {
      this.openNode(import_text_core.MarkupNodeType.paragraph);
      const node2 = this.marks.length > 0 ? { type: import_text_core.MarkupNodeType.text, text, marks: [...this.marks] } : { type: import_text_core.MarkupNodeType.text, text };
      this.push(node2);
      this.closeNode(import_text_core.MarkupNodeType.paragraph);
      return;
    }
    const node = this.marks.length > 0 ? { type: import_text_core.MarkupNodeType.text, text, marks: [...this.marks] } : { type: import_text_core.MarkupNodeType.text, text };
    this.push(node);
  }
  openMark(mark, attrs) {
    this.marks.push(attrs !== void 0 ? { type: mark, attrs } : { type: mark });
  }
  closeMark(mark) {
    if (this.marks[this.marks.length - 1]?.type === mark) {
      this.marks.pop();
    }
  }
  openNode(node, attrs) {
    this.stack.push(attrs !== void 0 ? { type: node, attrs } : { type: node });
  }
  closeNode(node) {
    this.marks.splice(0);
    const info = this.stack.pop();
    if (info !== void 0) {
      this.push(info);
    }
  }
  addNode(node) {
    this.push(node);
  }
  push(node) {
    const parent = this.top();
    if (parent !== void 0) {
      const content = parent.content ?? [];
      content.push(node);
      parent.content = content;
    }
  }
}
function nodeHandler({ node, getAttrs, wrapContent, wrapNode }) {
  const wrapStack = [];
  return {
    handleOpenTag: /* @__PURE__ */ __name((state, tag, attributes) => {
      const attrs = typeof getAttrs === "function" ? getAttrs(attributes) : typeof getAttrs === "object" ? { ...getAttrs } : void 0;
      const shouldWrapNode = wrapNode === true && state.top()?.type !== import_text_core.MarkupNodeType.paragraph;
      if (shouldWrapNode) {
        state.openNode(import_text_core.MarkupNodeType.paragraph);
      }
      wrapStack.push(shouldWrapNode);
      state.openNode(node, attrs);
      if (wrapContent === true) {
        state.openNode(import_text_core.MarkupNodeType.paragraph);
      }
    }, "handleOpenTag"),
    handleCloseTag: /* @__PURE__ */ __name((state, tag) => {
      if (wrapContent === true && state.top()?.type === import_text_core.MarkupNodeType.paragraph) {
        state.closeNode(import_text_core.MarkupNodeType.paragraph);
      }
      state.closeNode(node);
      if (wrapStack.pop() === true) {
        state.closeNode(import_text_core.MarkupNodeType.paragraph);
      }
    }, "handleCloseTag")
  };
}
__name(nodeHandler, "nodeHandler");
function markHandler({ mark, getAttrs }) {
  return {
    handleOpenTag: /* @__PURE__ */ __name((state, tag, attributes) => {
      const attrs = typeof getAttrs === "function" ? getAttrs(attributes) : typeof getAttrs === "object" ? { ...getAttrs } : void 0;
      state.openMark(mark, attrs);
    }, "handleOpenTag"),
    handleCloseTag: /* @__PURE__ */ __name((state, tag) => {
      state.closeMark(mark);
    }, "handleCloseTag")
  };
}
__name(markHandler, "markHandler");
function ignoreHandler(rule) {
  return {
    handleOpenTag: /* @__PURE__ */ __name(() => {
    }, "handleOpenTag"),
    handleCloseTag: /* @__PURE__ */ __name(() => {
    }, "handleCloseTag")
  };
}
__name(ignoreHandler, "ignoreHandler");
function specialHandler(rule) {
  return {
    handleOpenTag: /* @__PURE__ */ __name((state, tag, attributes) => {
      rule.handleOpenTag(state, tag, attributes);
    }, "handleOpenTag"),
    handleCloseTag: /* @__PURE__ */ __name((state, tag) => {
      rule.handleCloseTag(state, tag);
    }, "handleCloseTag")
  };
}
__name(specialHandler, "specialHandler");
const styleRules = [
  {
    style: "text-align",
    getAttrs: /* @__PURE__ */ __name((value) => {
      return { textAlign: value ?? null };
    }, "getAttrs")
  }
];
const markRules = {
  b: {
    mark: import_text_core.MarkupMarkType.bold
  },
  em: {
    mark: import_text_core.MarkupMarkType.em
  },
  i: {
    mark: import_text_core.MarkupMarkType.em
  },
  s: {
    mark: import_text_core.MarkupMarkType.strike
  },
  strong: {
    mark: import_text_core.MarkupMarkType.bold
  },
  u: {
    mark: import_text_core.MarkupMarkType.underline
  }
};
const nodeRules = {
  h1: {
    node: import_text_core.MarkupNodeType.heading,
    getAttrs: /* @__PURE__ */ __name((attributes) => {
      return {
        level: 1,
        textAlign: attributes.textAlign ?? null
      };
    }, "getAttrs")
  },
  h2: {
    node: import_text_core.MarkupNodeType.heading,
    getAttrs: /* @__PURE__ */ __name((attributes) => {
      return {
        level: 2,
        textAlign: attributes.textAlign ?? null
      };
    }, "getAttrs")
  },
  h3: {
    node: import_text_core.MarkupNodeType.heading,
    getAttrs: /* @__PURE__ */ __name((attributes) => {
      return {
        level: 3,
        textAlign: attributes.textAlign ?? null
      };
    }, "getAttrs")
  },
  h4: {
    node: import_text_core.MarkupNodeType.heading,
    getAttrs: /* @__PURE__ */ __name((attributes) => {
      return {
        level: 4,
        textAlign: attributes.textAlign ?? null
      };
    }, "getAttrs")
  },
  h5: {
    node: import_text_core.MarkupNodeType.heading,
    getAttrs: /* @__PURE__ */ __name((attributes) => {
      return {
        level: 5,
        textAlign: attributes.textAlign ?? null
      };
    }, "getAttrs")
  },
  h6: {
    node: import_text_core.MarkupNodeType.heading,
    getAttrs: /* @__PURE__ */ __name((attributes) => {
      return {
        level: 6,
        textAlign: attributes.textAlign ?? null
      };
    }, "getAttrs")
  },
  blockquote: {
    node: import_text_core.MarkupNodeType.blockquote
  },
  pre: {
    node: import_text_core.MarkupNodeType.code_block
  },
  hr: {
    node: import_text_core.MarkupNodeType.horizontal_rule
  },
  br: {
    node: import_text_core.MarkupNodeType.hard_break
  },
  ol: {
    node: import_text_core.MarkupNodeType.ordered_list,
    getAttrs: /* @__PURE__ */ __name((attributes) => {
      return {
        start: parseInt(attributes.starts ?? "1")
      };
    }, "getAttrs")
  },
  ul: {
    node: import_text_core.MarkupNodeType.bullet_list
  },
  li: {
    node: import_text_core.MarkupNodeType.list_item
  },
  img: {
    node: import_text_core.MarkupNodeType.image,
    wrapNode: true,
    getAttrs: /* @__PURE__ */ __name((attributes) => {
      return {
        src: attributes.src,
        alt: attributes.alt,
        title: attributes.title ?? null,
        width: attributes.width !== void 0 ? parseInt(attributes.width) : void 0,
        height: attributes.height !== void 0 ? parseInt(attributes.height) : void 0,
        "file-id": attributes["file-id"] ?? null
      };
    }, "getAttrs")
  },
  sub: {
    node: import_text_core.MarkupNodeType.subLink
  },
  table: {
    node: import_text_core.MarkupNodeType.table
  },
  tr: {
    node: import_text_core.MarkupNodeType.table_row
  },
  th: {
    node: import_text_core.MarkupNodeType.table_header,
    getAttrs: /* @__PURE__ */ __name((attributes) => {
      return {
        colspan: attributes.colspan !== void 0 ? parseInt(attributes.colspan) : void 0,
        rowspan: attributes.rowspan !== void 0 ? parseInt(attributes.rowspan) : void 0,
        colwidth: attributes.colwidth !== void 0 ? parseInt(attributes.colwidth) : void 0
      };
    }, "getAttrs"),
    wrapContent: true
  },
  td: {
    node: import_text_core.MarkupNodeType.table_cell,
    getAttrs: /* @__PURE__ */ __name((attributes) => {
      return {
        colspan: attributes.colspan !== void 0 ? parseInt(attributes.colspan) : void 0,
        rowspan: attributes.rowspan !== void 0 ? parseInt(attributes.rowspan) : void 0,
        colwidth: attributes.colwidth !== void 0 ? parseInt(attributes.colwidth) : void 0
      };
    }, "getAttrs"),
    wrapContent: true
  },
  comment: {
    node: import_text_core.MarkupNodeType.comment
  }
};
const specialRules = {
  p: {
    handleOpenTag: /* @__PURE__ */ __name((state, tag, attributes) => {
      const top = state.top();
      if (top?.type !== import_text_core.MarkupNodeType.paragraph) {
        const attrs = { textAlign: attributes.textAlign ?? null };
        state.openNode(import_text_core.MarkupNodeType.paragraph, attrs);
      }
    }, "handleOpenTag"),
    handleCloseTag: /* @__PURE__ */ __name((state, tag) => {
      const top = state.top();
      if (top?.type === import_text_core.MarkupNodeType.paragraph) {
        state.closeNode(import_text_core.MarkupNodeType.paragraph);
      }
    }, "handleCloseTag")
  },
  span: {
    handleOpenTag: /* @__PURE__ */ __name((state, tag, attributes) => {
      const dataType = attributes["data-type"];
      const dataColor = attributes["data-color"];
      const style = attributes.style;
      if (dataType === "reference") {
        state.openNode(import_text_core.MarkupNodeType.reference, {
          id: attributes["data-id"],
          objectclass: attributes["data-objectclass"],
          label: attributes["data-label"]
        });
      } else if (dataColor !== void 0) {
        state.openMark(import_text_core.MarkupMarkType.textColor, {
          color: dataColor
        });
      } else if (style !== void 0) {
        const attrs = {};
        style.split(";").forEach((part) => {
          const [key, value] = part.split(":");
          if (key !== void 0 && value !== void 0) {
            const trimmedKey = key.trim();
            const trimmedValue = value.trim();
            if (trimmedKey.length > 0 && trimmedValue.length > 0) {
              const camelKey = trimmedKey.replace(/-([a-z])/g, (g) => g[1].toUpperCase());
              attrs[camelKey] = trimmedValue;
            }
          }
        });
        if (Object.keys(attrs).length > 0) {
          state.openMark(import_text_core.MarkupMarkType.textStyle, attrs);
        }
      }
    }, "handleOpenTag"),
    handleCloseTag: /* @__PURE__ */ __name((state, tag) => {
      const top = state.top();
      if (top?.type === import_text_core.MarkupNodeType.reference) {
        delete top.content;
        state.closeNode(import_text_core.MarkupNodeType.reference);
      } else {
        state.closeMark(import_text_core.MarkupMarkType.textColor);
        state.closeMark(import_text_core.MarkupMarkType.textStyle);
      }
    }, "handleCloseTag")
  },
  code: {
    handleOpenTag: /* @__PURE__ */ __name((state, tag, attributes) => {
      const top = state.top();
      if (top?.type === import_text_core.MarkupNodeType.code_block) {
        const classes = attributes.class?.split(" ") ?? [];
        const language = classes.find((c) => c.startsWith("language-"));
        if (language !== void 0) {
          top.attrs = { language: language.substring(9) };
        }
      } else {
        state.openMark(import_text_core.MarkupMarkType.code);
      }
    }, "handleOpenTag"),
    handleCloseTag: /* @__PURE__ */ __name((state, tag) => {
      const top = state.top();
      if (top?.type === import_text_core.MarkupNodeType.code_block) {
      } else {
        state.closeMark(import_text_core.MarkupMarkType.code);
      }
    }, "handleCloseTag")
  },
  a: {
    handleOpenTag: /* @__PURE__ */ __name((state, tag, attributes) => {
      const dataType = attributes["data-type"];
      if (dataType === "embed") {
        state.openNode(import_text_core.MarkupNodeType.embed, {
          src: decodeURI(attributes.href)
        });
      } else {
        state.openMark(import_text_core.MarkupMarkType.link, {
          rel: attributes.rel,
          target: attributes.target,
          class: attributes.class,
          href: attributes.href,
          title: attributes.title
        });
      }
    }, "handleOpenTag"),
    handleCloseTag: /* @__PURE__ */ __name((state, tag) => {
      const top = state.top();
      if (top?.type === import_text_core.MarkupNodeType.embed) {
        delete top.content;
        state.closeNode(import_text_core.MarkupNodeType.embed);
      } else {
        state.closeMark(import_text_core.MarkupMarkType.link);
      }
    }, "handleCloseTag")
  }
};
const ignoreRules = {
  html: {},
  head: {},
  body: {},
  thead: {},
  tbody: {}
};
class HtmlParser {
  constructor(options = {}) {
    this.options = options;
    Object.entries(nodeRules).forEach(([tag, rule]) => {
      this.handlers[tag] = nodeHandler(rule);
    });
    Object.entries(markRules).forEach(([tag, rule]) => {
      this.handlers[tag] = markHandler(rule);
    });
    Object.entries(specialRules).forEach(([tag, rule]) => {
      this.handlers[tag] = specialHandler(rule);
    });
    Object.entries(ignoreRules).forEach(([tag, rule]) => {
      this.handlers[tag] = ignoreHandler(rule);
    });
  }
  static {
    __name(this, "HtmlParser");
  }
  handlers = {};
  parse(html) {
    const root = { type: import_text_core.MarkupNodeType.doc, content: [] };
    const state = new HtmlParseState(root, this.handlers);
    let rawPos;
    let rawDepth;
    const parser = new import_htmlparser2.Parser(
      {
        onopentag(tag, attributes) {
          if (rawDepth !== void 0) {
            rawDepth += 1;
            return;
          }
          if (attributes.style !== void 0) {
            const attrs = extractStyleAttrs(attributes, styleRules);
            if (attrs !== void 0) {
              attributes = { ...attributes, ...attrs };
            }
          }
          const handler = state.handlers[tag];
          if (handler !== void 0) {
            handler.handleOpenTag(state, tag, attributes);
          } else {
            rawPos = parser.startIndex;
            rawDepth = 0;
          }
        },
        ontext(text) {
          if (rawDepth !== void 0) {
            return;
          }
          if (text.length === 0) {
            return;
          }
          state.addText(text);
        },
        oncomment(text) {
          if (rawDepth !== void 0) {
            return;
          }
          state.openNode(import_text_core.MarkupNodeType.comment);
          state.addText(text);
          state.closeNode(import_text_core.MarkupNodeType.comment);
        },
        onclosetag(tag) {
          if (rawDepth !== void 0) {
            rawDepth -= 1;
            if (rawDepth === 0) {
              const start = rawPos ?? 0;
              const end = parser.endIndex;
              const text = html.substring(start, end);
              state.addText(text);
              rawPos = void 0;
              rawDepth = void 0;
            }
            return;
          }
          const handler = state.handlers[tag];
          if (handler !== void 0) {
            handler.handleCloseTag(state, tag);
          } else {
          }
        }
      },
      {
        decodeEntities: true,
        lowerCaseTags: true,
        recognizeSelfClosing: true
      }
    );
    parser.write(html);
    parser.end();
    return state.root;
  }
}
function extractStyleAttrs(attrs, rules) {
  const style = attrs.style;
  if (style !== void 0) {
    const styles = {};
    style.split(";").forEach((stylePart) => {
      const [key, value] = stylePart.split(":");
      styles[key.trim()] = value?.trim();
    });
    const result = {};
    rules.forEach((rule) => {
      if (styles[rule.style] !== void 0) {
        const attrs2 = rule.getAttrs?.(styles[rule.style]);
        if (attrs2 !== void 0) {
          Object.assign(result, attrs2);
        }
      }
    });
    return result;
  }
}
__name(extractStyleAttrs, "extractStyleAttrs");
