// @hcengineering/text-html/lib/serializer.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var serializer_exports = {};
__export(serializer_exports, {
  HtmlSerializer: () => HtmlSerializer
});
module.exports = __toCommonJS(serializer_exports);
var import_text_core = require("@hcengineering/text-core");
class HtmlSerializer {
  constructor(options = {}) {
    this.options = options;
  }
  static {
    __name(this, "HtmlSerializer");
  }
  serialize(markup) {
    const builder = new NodeBuilder(true);
    addNode(builder, markup);
    return builder.toText();
  }
}
class NodeBuilder {
  constructor(addTags) {
    this.addTags = addTags;
  }
  static {
    __name(this, "NodeBuilder");
  }
  textParts = [];
  addText(text) {
    this.textParts.push(text);
  }
  openTag(tag, attributes = {}, options) {
    if (this.addTags) {
      this.textParts.push("<");
      this.textParts.push(tag);
      for (const [key, value] of Object.entries(attributes)) {
        if (value == null) continue;
        if (typeof value === "boolean") {
          if (value) {
            this.textParts.push(` ${key}`);
          }
        } else {
          this.textParts.push(` ${key}="${escapeHtml(String(value))}"`);
        }
      }
      this.textParts.push(options?.selfClosing === true ? "/>" : ">");
    } else {
      if (options?.newLine === true) {
        this.textParts.push("\n");
      }
    }
  }
  closeTag(tag, options) {
    if (this.addTags) {
      this.textParts.push(`</${tag}>`);
    } else if (options?.newLine === true) {
      this.textParts.push("\n");
    }
  }
  toText() {
    return this.textParts.join("");
  }
}
function escapeHtml(text) {
  return text.replace(/&/g, "&amp;").replace(/</g, "&lt;").replace(/>/g, "&gt;").replace(/"/g, "&quot;").replace(/'/g, "&#039;").replace(/\r/g, "&#13;").replace(/\n/g, "&#10;");
}
__name(escapeHtml, "escapeHtml");
function addMark(builder, mark, next) {
  if (mark != null) {
    const attrs = mark.attrs ?? {};
    if (mark.type === import_text_core.MarkupMarkType.bold) {
      builder.openTag("strong");
      next?.();
      builder.closeTag("strong");
    } else if (mark.type === import_text_core.MarkupMarkType.code) {
      builder.openTag("code");
      next?.();
      builder.closeTag("code");
    } else if (mark.type === import_text_core.MarkupMarkType.em) {
      builder.openTag("em");
      next?.();
      builder.closeTag("em");
    } else if (mark.type === import_text_core.MarkupMarkType.link) {
      builder.openTag("a", {
        target: attrs.target,
        rel: attrs.rel,
        class: attrs.class,
        href: attrs.href,
        title: attrs.title
      });
      next?.();
      builder.closeTag("a");
    } else if (mark.type === import_text_core.MarkupMarkType.strike) {
      builder.openTag("s");
      next?.();
      builder.closeTag("s");
    } else if (mark.type === import_text_core.MarkupMarkType.underline) {
      builder.openTag("u");
      next?.();
      builder.closeTag("u");
    } else if (mark.type === import_text_core.MarkupMarkType.textColor) {
      const color = attrs.color;
      builder.openTag("span", {
        style: color !== void 0 ? `color: ${color}` : void 0,
        "data-color": color
      });
      next?.();
      builder.closeTag("span");
    } else if (mark.type === import_text_core.MarkupMarkType.textStyle) {
      const styleAttrs = Object.entries(attrs).map(([key, value]) => {
        const kebabKey = key.replace(/[A-Z]/g, (letter) => `-${letter.toLowerCase()}`);
        return `${kebabKey}: ${value}`;
      }).join("; ");
      builder.openTag("span", {
        style: styleAttrs
      });
      next?.();
      builder.closeTag("span");
    } else {
      builder.openTag("span", { "data-mark-type": mark.type });
      next?.();
      builder.closeTag("span");
    }
  }
}
__name(addMark, "addMark");
function addMarks(builder, marks, next) {
  if (marks.length > 0) {
    const mark = marks[0];
    const others = marks.slice(1);
    if (others.length > 0) {
      addMark(builder, mark, () => {
        addMarks(builder, others, next);
      });
    } else {
      addMark(builder, mark, next);
    }
  }
}
__name(addMarks, "addMarks");
function addNodes(builder, nodes) {
  nodes.forEach((childNode) => {
    addNode(builder, childNode);
  });
}
__name(addNodes, "addNodes");
function addNodeContent(builder, node) {
  if (node == null) return;
  const attrs = node.attrs ?? {};
  const nodes = node.content ?? [];
  const style = toStyleAttr(attrs);
  if (node.type === import_text_core.MarkupNodeType.doc) {
    addNodes(builder, nodes);
  } else if (node.type === import_text_core.MarkupNodeType.paragraph) {
    builder.openTag("p", { style });
    addNodes(builder, nodes);
    builder.closeTag("p");
  } else if (node.type === import_text_core.MarkupNodeType.blockquote) {
    builder.openTag("blockquote");
    addNodes(builder, nodes);
    builder.closeTag("blockquote");
  } else if (node.type === import_text_core.MarkupNodeType.horizontal_rule) {
    builder.openTag("hr", {}, { selfClosing: true });
  } else if (node.type === import_text_core.MarkupNodeType.heading) {
    const level = toNumber(attrs.level) ?? 1;
    const tag = `h${level}`;
    builder.openTag(tag, { style });
    addNodes(builder, nodes);
    builder.closeTag(tag);
  } else if (node.type === import_text_core.MarkupNodeType.code_block) {
    const attrs2 = node.attrs?.language !== void 0 ? { class: `language-${node.attrs.language}` } : {};
    builder.openTag("pre");
    builder.openTag("code", attrs2);
    addNodes(builder, nodes);
    builder.closeTag("code");
    builder.closeTag("pre");
  } else if (node.type === import_text_core.MarkupNodeType.text) {
    builder.addText(node.text ?? "");
  } else if (node.type === import_text_core.MarkupNodeType.image) {
    const src = toString(attrs.src);
    const alt = toString(attrs.alt);
    const width = toString(attrs.width);
    const height = toString(attrs.height);
    builder.openTag("img", { src, alt, width, height }, { selfClosing: true });
  } else if (node.type === import_text_core.MarkupNodeType.reference) {
    const label = toString(attrs.label);
    builder.openTag("span", {
      "data-type": "reference",
      "data-id": attrs.id,
      "data-objectclass": attrs.objectclass,
      "data-label": attrs.label
    });
    builder.addText(label !== void 0 ? `@${label}` : "");
    builder.closeTag("span");
  } else if (node.type === import_text_core.MarkupNodeType.hard_break) {
    builder.openTag("br", {}, { selfClosing: true });
  } else if (node.type === import_text_core.MarkupNodeType.ordered_list) {
    builder.openTag("ol");
    addNodes(builder, nodes);
    builder.closeTag("ol");
  } else if (node.type === import_text_core.MarkupNodeType.bullet_list) {
    builder.openTag("ul");
    addNodes(builder, nodes);
    builder.closeTag("ul");
  } else if (node.type === import_text_core.MarkupNodeType.list_item) {
    builder.openTag("li");
    addNodes(builder, nodes);
    builder.closeTag("li");
  } else if (node.type === import_text_core.MarkupNodeType.todoList) {
    builder.openTag("ul", { "data-type": import_text_core.MarkupNodeType.todoList });
    addNodes(builder, nodes);
    builder.closeTag("ul");
  } else if (node.type === import_text_core.MarkupNodeType.todoItem) {
    const checked = node.attrs?.checked === true || node.attrs?.checked === "true";
    const disabled = node.attrs?.disabled === true || node.attrs?.disabled === "true";
    builder.openTag("li", {
      "data-type": import_text_core.MarkupNodeType.todoItem,
      "data-todoid": node.attrs?.todoid,
      "data-userid": node.attrs?.userid,
      "data-checked": checked
    });
    builder.openTag("input", { type: "checkbox", checked, disabled }, { selfClosing: true });
    addNodes(builder, nodes);
    builder.closeTag("li");
  } else if (node.type === import_text_core.MarkupNodeType.taskList) {
    builder.openTag("ul", { "data-type": import_text_core.MarkupNodeType.taskList });
    addNodes(builder, nodes);
    builder.closeTag("ul");
  } else if (node.type === import_text_core.MarkupNodeType.taskItem) {
    const checked = node.attrs?.checked === true || node.attrs?.checked === "true";
    const disabled = node.attrs?.disabled === true || node.attrs?.disabled === "true";
    builder.openTag("li", {
      "data-type": import_text_core.MarkupNodeType.taskItem,
      "data-checked": checked
    });
    builder.openTag("input", { type: "checkbox", checked, disabled }, { selfClosing: true });
    addNodes(builder, nodes);
    builder.closeTag("li");
  } else if (node.type === import_text_core.MarkupNodeType.subLink) {
    builder.openTag("sub");
    addNodes(builder, nodes);
    builder.closeTag("sub");
  } else if (node.type === import_text_core.MarkupNodeType.table) {
    builder.openTag("table");
    builder.openTag("tbody");
    addNodes(builder, nodes);
    builder.closeTag("tbody");
    builder.closeTag("table");
  } else if (node.type === import_text_core.MarkupNodeType.table_row) {
    builder.openTag("tr");
    addNodes(builder, nodes);
    builder.closeTag("tr");
  } else if (node.type === import_text_core.MarkupNodeType.table_cell) {
    const colspan = toNumber(attrs.colspan) ?? 1;
    const rowspan = toNumber(attrs.rowspan) ?? 1;
    const colwidth = toNumber(attrs.colwidth);
    builder.openTag("td", {
      colspan: colspan !== 1 ? colspan : void 0,
      rowspan: rowspan !== 1 ? rowspan : void 0,
      colwidth: colwidth !== void 0 && colwidth > 0 ? colwidth : void 0
    });
    addNodes(builder, nodes);
    builder.closeTag("td");
  } else if (node.type === import_text_core.MarkupNodeType.table_header) {
    const colspan = toNumber(attrs.colspan) ?? 1;
    const rowspan = toNumber(attrs.rowspan) ?? 1;
    const colwidth = toNumber(attrs.colwidth);
    builder.openTag("th", {
      colspan: colspan !== 1 ? colspan : void 0,
      rowspan: rowspan !== 1 ? rowspan : void 0,
      colwidth: colwidth !== void 0 && colwidth > 0 ? colwidth : void 0
    });
    addNodes(builder, nodes);
    builder.closeTag("th");
  } else if (node.type === import_text_core.MarkupNodeType.comment) {
    builder.addText("<!-- ");
    addNodes(builder, nodes);
    builder.addText(" -->");
  } else if (node.type === import_text_core.MarkupNodeType.embed) {
    const src = toString(attrs.src) ?? "";
    builder.openTag("a", { href: encodeURI(src), "data-type": "embed" });
    builder.addText(escapeHtml(src));
    builder.closeTag("a");
  } else {
    builder.openTag("div", { "data-node-type": node.type });
    addNodes(builder, nodes);
    builder.closeTag("div");
  }
}
__name(addNodeContent, "addNodeContent");
function addNode(builder, node) {
  const marks = node.marks ?? [];
  if (marks.length > 0) {
    addMarks(builder, marks, () => {
      addNodeContent(builder, node);
    });
  } else {
    addNodeContent(builder, node);
  }
}
__name(addNode, "addNode");
function toString(value) {
  return value !== void 0 ? `${value}` : void 0;
}
__name(toString, "toString");
function toNumber(value) {
  if (typeof value === "boolean") {
    return value ? 1 : 0;
  }
  return value != null ? typeof value === "string" ? parseInt(value) : value : void 0;
}
__name(toNumber, "toNumber");
function toStyleAttr(attrs) {
  const styles = [];
  if (attrs.textAlign != null) {
    styles.push(`text-align: ${attrs.textAlign}`);
  }
  return styles.length > 0 ? styles.join("; ") : void 0;
}
__name(toStyleAttr, "toStyleAttr");
