// @hcengineering/client/lib/__tests__/client.test.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var import_core = __toESM(require("@hcengineering/core"));
const txFactory = new import_core.TxFactory(import_core.default.account.System);
function createClass(_class, attributes) {
  return txFactory.createTxCreateDoc(import_core.default.class.Class, import_core.default.space.Model, attributes, _class);
}
__name(createClass, "createClass");
function generateMinimalModel() {
  const txes = [];
  txes.push(createClass(import_core.default.class.Obj, { label: "Obj", kind: import_core.ClassifierKind.CLASS }));
  txes.push(
    createClass(import_core.default.class.Doc, { label: "Doc", extends: import_core.default.class.Obj, kind: import_core.ClassifierKind.CLASS })
  );
  txes.push(
    createClass(import_core.default.class.Class, {
      label: "Class",
      extends: import_core.default.class.Doc,
      kind: import_core.ClassifierKind.CLASS,
      domain: import_core.DOMAIN_MODEL
    })
  );
  txes.push(
    createClass(import_core.default.class.Space, {
      label: "Space",
      extends: import_core.default.class.Doc,
      kind: import_core.ClassifierKind.CLASS,
      domain: import_core.DOMAIN_MODEL
    })
  );
  txes.push(
    createClass(import_core.default.class.Tx, {
      label: "Tx",
      extends: import_core.default.class.Doc,
      kind: import_core.ClassifierKind.CLASS,
      domain: import_core.DOMAIN_TX
    })
  );
  txes.push(
    createClass(import_core.default.class.TxCUD, {
      label: "TxCUD",
      extends: import_core.default.class.Tx,
      kind: import_core.ClassifierKind.CLASS,
      domain: import_core.DOMAIN_TX
    })
  );
  txes.push(
    createClass(import_core.default.class.TxCreateDoc, {
      label: "TxCreateDoc",
      extends: import_core.default.class.TxCUD,
      kind: import_core.ClassifierKind.CLASS
    })
  );
  txes.push(
    createClass(import_core.default.class.TxUpdateDoc, {
      label: "TxUpdateDoc",
      extends: import_core.default.class.TxCUD,
      kind: import_core.ClassifierKind.CLASS
    })
  );
  txes.push(
    createClass(import_core.default.class.TxRemoveDoc, {
      label: "TxRemoveDoc",
      extends: import_core.default.class.TxCUD,
      kind: import_core.ClassifierKind.CLASS
    })
  );
  return txes;
}
__name(generateMinimalModel, "generateMinimalModel");
class TestConnection {
  static {
    __name(this, "TestConnection");
  }
  hierarchy;
  model;
  transactions;
  handlers = [];
  _connected = true;
  onConnect;
  getLastHash;
  constructor(txes) {
    this.hierarchy = new import_core.Hierarchy();
    for (const tx of txes) {
      this.hierarchy.tx(tx);
    }
    this.transactions = new import_core.TxDb(this.hierarchy);
    this.model = new import_core.ModelDb(this.hierarchy);
    for (const tx of txes) {
      void this.transactions.tx(tx);
      void this.model.tx(tx);
    }
  }
  isConnected() {
    return this._connected;
  }
  setConnected(value) {
    this._connected = value;
  }
  pushHandler(handler) {
    this.handlers.push(handler);
  }
  async findAll(_class, query, options) {
    const domain = this.hierarchy.getClass(_class).domain;
    if (domain === import_core.DOMAIN_TX) {
      return await this.transactions.findAll(_class, query, options);
    }
    return await this.model.findAll(_class, query, options);
  }
  async searchFulltext(query, options) {
    return { docs: [] };
  }
  async tx(tx) {
    if (tx.objectSpace === import_core.default.space.Model) {
      this.hierarchy.tx(tx);
      await this.model.tx(tx);
    }
    await this.transactions.tx(tx);
    this.handlers.forEach((h) => {
      h(tx);
    });
    return {};
  }
  async loadModel(last, hash) {
    const txes = await this.transactions.findAll(import_core.default.class.Tx, {
      objectSpace: import_core.default.space.Model,
      modifiedOn: { $gt: last }
    });
    if (hash !== void 0) {
      return {
        transactions: txes,
        hash: "test-hash",
        full: false
      };
    }
    return txes;
  }
  async close() {
    this._connected = false;
  }
  async loadChunk(domain, idx) {
    return {
      idx: idx ?? 0,
      docs: [],
      finished: true
    };
  }
  async getDomainHash(domain) {
    return "test-hash";
  }
  async closeChunk(idx) {
  }
  async loadDocs(domain, docs) {
    return [];
  }
  async upload(domain, docs) {
  }
  async clean(domain, docs) {
  }
  async sendForceClose() {
  }
  async domainRequest(ctx, params, options) {
    return { domain: ctx, value: null };
  }
  // Simulate receiving transactions from server
  simulateTransaction(tx) {
    this.handlers.forEach((h) => {
      h(tx);
    });
  }
}
describe("Client Core Implementation", () => {
  let testConnection;
  let client;
  beforeEach(async () => {
    const txes = generateMinimalModel();
    testConnection = new TestConnection(txes);
    client = await (0, import_core.createClient)(async (handler) => {
      testConnection.pushHandler(handler);
      return testConnection;
    });
  });
  it("should create a client instance", () => {
    expect(client).toBeDefined();
    expect(client.getHierarchy()).toBeDefined();
    expect(client.getModel()).toBeDefined();
  });
  it("should handle findAll operations", async () => {
    const spaces = await client.findAll(import_core.default.class.Space, {});
    expect(spaces).toBeDefined();
    expect(Array.isArray(spaces)).toBe(true);
  });
  it("should handle findOne operations", async () => {
    const space = await client.findOne(import_core.default.class.Space, { name: "TestSpace" });
    if (space !== void 0) {
      expect(space.name).toBe("TestSpace");
    }
  });
  it("should handle tx operations", async () => {
    const tx = {
      _id: (0, import_core.generateId)(),
      _class: import_core.default.class.TxCreateDoc,
      space: import_core.default.space.Tx,
      objectId: (0, import_core.generateId)(),
      objectClass: import_core.default.class.Space,
      objectSpace: import_core.default.space.Model,
      modifiedBy: import_core.default.account.System,
      modifiedOn: Date.now(),
      createdOn: Date.now(),
      attributes: {
        name: "NewSpace",
        description: "",
        private: false,
        archived: false,
        members: []
      }
    };
    const result = await client.tx(tx);
    expect(result).toBeDefined();
  });
  it("should handle updateFromRemote for model transactions", async () => {
    const notifySpy = jest.fn();
    client.notify = notifySpy;
    const tx = {
      _id: (0, import_core.generateId)(),
      _class: import_core.default.class.TxCreateDoc,
      space: import_core.default.space.Tx,
      objectId: (0, import_core.generateId)(),
      objectClass: import_core.default.class.Space,
      objectSpace: import_core.default.space.Model,
      modifiedBy: import_core.default.account.System,
      modifiedOn: Date.now(),
      createdOn: Date.now(),
      attributes: {
        name: "RemoteSpace",
        description: "",
        private: false,
        archived: false,
        members: []
      }
    };
    testConnection.simulateTransaction(tx);
    await new Promise((resolve) => setTimeout(resolve, 100));
    expect(notifySpy).toHaveBeenCalled();
  });
  it("should close connection properly", async () => {
    await client.close();
    expect(testConnection.isConnected()).toBe(false);
  });
  it("should handle domainRequest", async () => {
    const result = await client.domainRequest("test-domain", {});
    expect(result).toBeDefined();
  });
  it("should update hierarchy and model on model transactions", async () => {
    const hierarchy = client.getHierarchy();
    const model = client.getModel();
    expect(hierarchy).toBeDefined();
    expect(model).toBeDefined();
    const tx = {
      _id: (0, import_core.generateId)(),
      _class: import_core.default.class.TxCreateDoc,
      space: import_core.default.space.Tx,
      objectId: (0, import_core.generateId)(),
      objectClass: import_core.default.class.Space,
      objectSpace: import_core.default.space.Model,
      modifiedBy: import_core.default.account.System,
      modifiedOn: Date.now(),
      createdOn: Date.now(),
      attributes: {
        name: "ModelSpace",
        description: "",
        private: false,
        archived: false,
        members: []
      }
    };
    await client.tx(tx);
    const spaces = await client.findAll(import_core.default.class.Space, { name: "ModelSpace" });
    expect(spaces.length).toBeGreaterThanOrEqual(0);
  });
  it("should handle reconnection events", async () => {
    let eventReceived;
    const onConnectHandler = /* @__PURE__ */ __name(async (event, lastTx, data) => {
      eventReceived = event;
    }, "onConnectHandler");
    testConnection.onConnect = onConnectHandler;
    if (testConnection.onConnect !== void 0) {
      await testConnection.onConnect(import_core.ClientConnectEvent.Connected, void 0, {});
    }
    expect(eventReceived).toBe(import_core.ClientConnectEvent.Connected);
  });
  it("should handle workspace events", async () => {
    const notifySpy = jest.fn();
    client.notify = notifySpy;
    const workspaceTx = {
      _id: (0, import_core.generateId)(),
      _class: import_core.default.class.TxWorkspaceEvent,
      space: import_core.default.space.Tx,
      objectSpace: import_core.default.space.Workspace,
      modifiedBy: import_core.default.account.System,
      modifiedOn: Date.now(),
      createdOn: Date.now(),
      event: import_core.WorkspaceEvent.LastTx,
      params: { lastTx: "test-last-tx" }
    };
    testConnection.simulateTransaction(workspaceTx);
    await new Promise((resolve) => setTimeout(resolve, 100));
    expect(notifySpy).toHaveBeenCalled();
  });
  it("should handle searchFulltext", async () => {
    const query = { query: "test" };
    const options = {};
    const result = await client.searchFulltext(query, options);
    expect(result).toBeDefined();
    expect(result.docs).toBeDefined();
  });
  it("should handle mixin updates in findAll", async () => {
    const spaces = await client.findAll(import_core.default.class.Space, {}, { limit: 10 });
    expect(spaces).toBeDefined();
    expect(spaces.total).toBeDefined();
  });
  it("should skip already applied model transactions", async () => {
    const tx = {
      _id: (0, import_core.generateId)(),
      _class: import_core.default.class.TxCreateDoc,
      space: import_core.default.space.Tx,
      objectId: (0, import_core.generateId)(),
      objectClass: import_core.default.class.Space,
      objectSpace: import_core.default.space.Model,
      modifiedBy: import_core.default.account.System,
      modifiedOn: Date.now(),
      createdOn: Date.now(),
      attributes: {
        name: "DuplicateSpace",
        description: "",
        private: false,
        archived: false,
        members: []
      }
    };
    await client.tx(tx);
    const notifySpy = jest.fn();
    client.notify = notifySpy;
    testConnection.simulateTransaction(tx);
    await new Promise((resolve) => setTimeout(resolve, 100));
    expect(notifySpy).toHaveBeenCalled();
  });
});
