// @hcengineering/api-client/lib/storage/client.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var client_exports = {};
__export(client_exports, {
  StorageClientImpl: () => StorageClientImpl,
  connectStorage: () => connectStorage,
  createStorageClient: () => createStorageClient
});
module.exports = __toCommonJS(client_exports);
var import_core = __toESM(require("@hcengineering/core"));
var import_stream = require("stream");
var import_config = require("../config");
var import_error = require("./error");
var import_utils = require("../utils");
class StorageClientImpl {
  constructor(filesUrl, uploadUrl, token, workspace) {
    this.filesUrl = filesUrl;
    this.uploadUrl = uploadUrl;
    this.workspace = workspace;
    this.headers = {
      Authorization: "Bearer " + token
    };
  }
  static {
    __name(this, "StorageClientImpl");
  }
  headers;
  getObjectUrl(objectName) {
    return this.filesUrl.replace(":filename", objectName).replace(":blobId", objectName);
  }
  async stat(objectName) {
    const url = this.getObjectUrl(objectName);
    let response;
    try {
      response = await wrappedFetch(url, { method: "HEAD", headers: { ...this.headers } });
    } catch (error) {
      if (error instanceof import_error.NotFoundError) {
        return;
      }
      throw error;
    }
    const headers = response.headers;
    const lastModified = Date.parse(headers.get("Last-Modified") ?? "");
    const size = parseInt(headers.get("Content-Length") ?? "0", 10);
    return {
      provider: "",
      _class: import_core.default.class.Blob,
      _id: objectName,
      contentType: headers.get("Content-Type") ?? "",
      size: isNaN(size) ? 0 : size ?? 0,
      etag: headers.get("ETag") ?? "",
      space: import_core.default.space.Configuration,
      modifiedBy: import_core.default.account.System,
      modifiedOn: isNaN(lastModified) ? 0 : lastModified,
      version: null
    };
  }
  async get(objectName) {
    const url = this.getObjectUrl(objectName);
    const response = await wrappedFetch(url, { headers: { ...this.headers } });
    if (response.body == null) {
      throw new import_error.StorageError("Missing response body");
    }
    return import_stream.Readable.from(response.body);
  }
  async put(objectName, stream, contentType, size) {
    const buffer = await toBuffer(stream);
    const file = new File([new Uint8Array(buffer)], objectName, { type: contentType });
    const formData = new FormData();
    formData.append("file", file);
    let response;
    try {
      response = await fetch(this.uploadUrl, {
        method: "POST",
        body: formData,
        headers: { ...this.headers }
      });
    } catch (error) {
      throw new import_error.NetworkError(`Network error ${error}`);
    }
    if (!response.ok) {
      throw new import_error.StorageError(await response.text());
    }
    const result = await response.json();
    if (Object.hasOwn(result[0], "id")) {
      const fileResult = result[0];
      return {
        _class: import_core.default.class.Blob,
        _id: fileResult.id,
        space: import_core.default.space.Configuration,
        modifiedOn: fileResult.metadata.lastModified,
        modifiedBy: import_core.default.account.System,
        provider: "",
        contentType: fileResult.metadata.contentType,
        etag: fileResult.metadata.etag,
        version: null,
        size: fileResult.metadata.size
      };
    } else {
      const error = result[0] ?? "Unknown error";
      throw new import_error.StorageError(`Storage error ${error.error}`);
    }
  }
  async partial(objectName, offset, length) {
    const url = this.getObjectUrl(objectName);
    const response = await wrappedFetch(url, {
      headers: {
        ...this.headers,
        Range: length !== void 0 ? `bytes=${offset}-${offset + length - 1}` : `bytes=${offset}`
      }
    });
    if (response.body == null) {
      throw new import_error.StorageError("Missing response body");
    }
    return import_stream.Readable.from(response.body);
  }
  async remove(objectName) {
    const url = this.getObjectUrl(objectName);
    await wrappedFetch(url, {
      method: "DELETE",
      headers: { ...this.headers }
    });
  }
}
async function toBuffer(data) {
  if (Buffer.isBuffer(data)) {
    return data;
  } else if (typeof data === "string") {
    return Buffer.from(data);
  } else if (data instanceof import_stream.Readable) {
    const chunks = [];
    for await (const chunk of data) {
      chunks.push(chunk);
    }
    return Buffer.concat(chunks);
  } else {
    throw new TypeError("Unsupported data type");
  }
}
__name(toBuffer, "toBuffer");
async function wrappedFetch(url, init) {
  let response;
  try {
    response = await fetch(url, init);
  } catch (error) {
    throw new import_error.NetworkError(`Network error ${error}`);
  }
  if (!response.ok) {
    const text = await response.text();
    if (response.status === 404) {
      throw new import_error.NotFoundError(text);
    } else {
      throw new import_error.StorageError(text);
    }
  }
  return response;
}
__name(wrappedFetch, "wrappedFetch");
function createStorageClient(filesUrl, uploadUrl, token, workspace) {
  return new StorageClientImpl(filesUrl, uploadUrl, token, workspace);
}
__name(createStorageClient, "createStorageClient");
async function connectStorage(url, options, config) {
  config ??= await (0, import_config.loadServerConfig)(url);
  const token = await (0, import_utils.getWorkspaceToken)(url, options, config);
  const filesUrl = (config.FILES_URL.startsWith("/") ? (0, import_core.concatLink)(url, config.FILES_URL) : config.FILES_URL).replace(
    ":workspace",
    token.workspaceId
  );
  const uploadUrl = (config.UPLOAD_URL.startsWith("/") ? (0, import_core.concatLink)(url, config.UPLOAD_URL) : config.UPLOAD_URL).replace(":workspace", token.workspaceId);
  return new StorageClientImpl(filesUrl, uploadUrl, token.token, token.workspaceId);
}
__name(connectStorage, "connectStorage");
