// @hcengineering/client-resources/lib/__tests__/connection.test.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var import_client = require("@hcengineering/client");
var import_core = __toESM(require("@hcengineering/core"));
var import_connection = require("../connection");
class MockCloseEvent {
  static {
    __name(this, "MockCloseEvent");
  }
  type;
  code;
  constructor(type, init) {
    this.type = type;
    this.code = init?.code ?? 1e3;
  }
}
class MockMessageEvent {
  static {
    __name(this, "MockMessageEvent");
  }
  type;
  data;
  constructor(type, init) {
    this.type = type;
    this.data = init.data;
  }
}
class MockEvent {
  static {
    __name(this, "MockEvent");
  }
  type;
  constructor(type) {
    this.type = type;
  }
}
class MockWebSocket {
  constructor(url) {
    this.url = url;
    const timer = setTimeout(() => {
      if (this.readyState === import_client.ClientSocketReadyState.CLOSED) {
        return;
      }
      this.readyState = import_client.ClientSocketReadyState.OPEN;
      if (this.onopen !== null && this.onopen !== void 0) {
        this.onopen(new MockEvent("open"));
      }
      this.processQueue();
    }, 10);
    this.timers.push(timer);
  }
  static {
    __name(this, "MockWebSocket");
  }
  readyState = import_client.ClientSocketReadyState.CONNECTING;
  onmessage = null;
  onclose = null;
  onopen = null;
  onerror = null;
  bufferedAmount = 0;
  messageQueue = [];
  closeCode;
  timers = [];
  clearAllTimers() {
    for (const timer of this.timers) {
      clearTimeout(timer);
    }
    this.timers = [];
  }
  send(data) {
    if (this.readyState !== import_client.ClientSocketReadyState.OPEN) {
      throw new Error("WebSocket is not open");
    }
    try {
      const message = typeof data === "string" ? data : new TextDecoder().decode(data);
      if (message === import_client.pingConst) {
        this.simulateMessage(import_client.pongConst);
        return;
      }
      const rpcMessage = JSON.parse(message);
      if (rpcMessage.method === "hello") {
        const response2 = {
          id: rpcMessage.id,
          result: {
            binary: false,
            compression: false,
            reconnect: false
          }
        };
        this.simulateMessage(JSON.stringify(response2));
        return;
      }
      if (rpcMessage.method === "loadModel") {
        const response2 = {
          id: rpcMessage.id,
          result: {
            transactions: [],
            hash: "test-hash",
            full: false
          }
        };
        this.simulateMessage(JSON.stringify(response2));
        return;
      }
      if (rpcMessage.method === "findAll") {
        const response2 = {
          id: rpcMessage.id,
          result: []
        };
        this.simulateMessage(JSON.stringify(response2));
        return;
      }
      if (rpcMessage.method === "tx") {
        const response2 = {
          id: rpcMessage.id,
          result: {}
        };
        this.simulateMessage(JSON.stringify(response2));
        return;
      }
      const response = {
        id: rpcMessage.id,
        result: {}
      };
      this.simulateMessage(JSON.stringify(response));
    } catch (err) {
      console.error("Error processing message:", err);
    }
  }
  close(code) {
    this.closeCode = code;
    this.readyState = import_client.ClientSocketReadyState.CLOSING;
    this.clearAllTimers();
    const timer = setTimeout(() => {
      this.readyState = import_client.ClientSocketReadyState.CLOSED;
      if (this.onclose !== null && this.onclose !== void 0) {
        this.onclose(new MockCloseEvent("close", { code: code ?? 1e3 }));
      }
    }, 10);
    this.timers.push(timer);
  }
  // Helper method to simulate receiving messages
  simulateMessage(data) {
    if (this.readyState === import_client.ClientSocketReadyState.OPEN) {
      if (this.onmessage !== null && this.onmessage !== void 0) {
        const event = new MockMessageEvent("message", { data });
        this.onmessage(event);
      }
    } else {
      this.messageQueue.push(data);
    }
  }
  processQueue() {
    while (this.messageQueue.length > 0) {
      const data = this.messageQueue.shift();
      this.simulateMessage(data);
    }
  }
  // Helper to simulate server-initiated transactions
  simulateTransaction(tx) {
    const message = {
      result: {
        _class: "core:class:TxNotification",
        tx: [tx]
      }
    };
    this.simulateMessage(JSON.stringify(message));
  }
}
describe("MockWebSocket", () => {
  it("should connect and send messages", async () => {
    const ws = new MockWebSocket("ws://localhost:3333");
    await new Promise((resolve) => {
      ws.onopen = () => {
        expect(ws.readyState).toBe(import_client.ClientSocketReadyState.OPEN);
        resolve();
      };
    });
    let receivedMessage = null;
    ws.onmessage = (ev) => {
      receivedMessage = ev.data;
    };
    ws.simulateMessage("test message");
    expect(receivedMessage).toBe("test message");
    await new Promise((resolve) => {
      ws.onclose = () => {
        expect(ws.readyState).toBe(import_client.ClientSocketReadyState.CLOSED);
        resolve();
      };
      ws.close();
    });
  });
  it("should handle ping/pong", async () => {
    const ws = new MockWebSocket("ws://localhost:3333");
    await new Promise((resolve) => {
      ws.onopen = () => {
        resolve();
      };
    });
    let receivedMessage = null;
    ws.onmessage = (ev) => {
      receivedMessage = ev.data;
    };
    ws.send(import_client.pingConst);
    await new Promise((resolve) => setTimeout(resolve, 50));
    expect(receivedMessage).toBe(import_client.pongConst);
    ws.close();
  });
});
describe("connect function", () => {
  let connections = [];
  let mockWebSockets = [];
  afterEach(async () => {
    for (const conn of connections) {
      await conn.close();
    }
    connections = [];
    for (const ws of mockWebSockets) {
      ws.clearAllTimers();
      if (ws.readyState !== import_client.ClientSocketReadyState.CLOSED) {
        ws.close();
      }
    }
    mockWebSockets = [];
    await new Promise((resolve) => setTimeout(resolve, 100));
  });
  it("should establish connection", async () => {
    const workspaceId = "test-workspace";
    const userId = "test-user";
    const handler = jest.fn();
    const mockWs = new MockWebSocket("ws://localhost:3333");
    mockWebSockets.push(mockWs);
    const client = (0, import_connection.connect)("ws://localhost:3333", handler, workspaceId, userId, {
      socketFactory: /* @__PURE__ */ __name((url) => mockWs, "socketFactory")
    });
    connections.push(client);
    expect(client).toBeDefined();
    await new Promise((resolve) => setTimeout(resolve, 100));
    await client.close();
    await new Promise((resolve) => setTimeout(resolve, 50));
  });
  it("should handle transactions", async () => {
    const workspaceId = "test-workspace";
    const userId = "test-user";
    let txReceived = null;
    const handler = /* @__PURE__ */ __name((...tx) => {
      if (tx.length > 0) {
        txReceived = tx[0];
      }
    }, "handler");
    const mockWs = new MockWebSocket("ws://localhost:3333");
    const client = (0, import_connection.connect)("ws://localhost:3333", handler, workspaceId, userId, {
      socketFactory: /* @__PURE__ */ __name((url) => mockWs, "socketFactory")
    });
    connections.push(client);
    await new Promise((resolve) => setTimeout(resolve, 150));
    const testTx = {
      _id: (0, import_core.generateId)(),
      _class: import_core.default.class.TxCreateDoc,
      space: import_core.default.space.Model,
      objectId: (0, import_core.generateId)(),
      objectClass: import_core.default.class.Space,
      objectSpace: import_core.default.space.Model,
      modifiedBy: import_core.default.account.System,
      modifiedOn: Date.now(),
      createdOn: Date.now(),
      attributes: {
        name: "Test Space",
        description: "",
        private: false,
        archived: false,
        members: []
      }
    };
    mockWs.simulateTransaction(testTx);
    await new Promise((resolve) => setTimeout(resolve, 100));
    expect(txReceived).toBeDefined();
    await client.close();
    await new Promise((resolve) => setTimeout(resolve, 50));
  });
});
