// @hcengineering/client-resources/lib/__tests__/integration.test.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var integration_test_exports = {};
__export(integration_test_exports, {
  MockClientConnection: () => MockClientConnection,
  createTestClient: () => createTestClient,
  createTestTx: () => createTestTx
});
module.exports = __toCommonJS(integration_test_exports);
var import_core = __toESM(require("@hcengineering/core"));
const txFactory = new import_core.TxFactory(import_core.default.account.System);
function createClass(_class, attributes) {
  return txFactory.createTxCreateDoc(import_core.default.class.Class, import_core.default.space.Model, attributes, _class);
}
__name(createClass, "createClass");
function generateMinimalModel() {
  const txes = [];
  txes.push(createClass(import_core.default.class.Obj, { label: "Obj", kind: import_core.ClassifierKind.CLASS }));
  txes.push(
    createClass(import_core.default.class.Doc, { label: "Doc", extends: import_core.default.class.Obj, kind: import_core.ClassifierKind.CLASS })
  );
  txes.push(
    createClass(import_core.default.class.Class, {
      label: "Class",
      extends: import_core.default.class.Doc,
      kind: import_core.ClassifierKind.CLASS,
      domain: import_core.DOMAIN_MODEL
    })
  );
  txes.push(
    createClass(import_core.default.class.Space, {
      label: "Space",
      extends: import_core.default.class.Doc,
      kind: import_core.ClassifierKind.CLASS,
      domain: import_core.DOMAIN_MODEL
    })
  );
  txes.push(
    createClass(import_core.default.class.Tx, {
      label: "Tx",
      extends: import_core.default.class.Doc,
      kind: import_core.ClassifierKind.CLASS,
      domain: import_core.DOMAIN_TX
    })
  );
  txes.push(
    createClass(import_core.default.class.TxCUD, {
      label: "TxCUD",
      extends: import_core.default.class.Tx,
      kind: import_core.ClassifierKind.CLASS,
      domain: import_core.DOMAIN_TX
    })
  );
  txes.push(
    createClass(import_core.default.class.TxCreateDoc, {
      label: "TxCreateDoc",
      extends: import_core.default.class.TxCUD,
      kind: import_core.ClassifierKind.CLASS
    })
  );
  txes.push(
    createClass(import_core.default.class.TxUpdateDoc, {
      label: "TxUpdateDoc",
      extends: import_core.default.class.TxCUD,
      kind: import_core.ClassifierKind.CLASS
    })
  );
  txes.push(
    createClass(import_core.default.class.TxRemoveDoc, {
      label: "TxRemoveDoc",
      extends: import_core.default.class.TxCUD,
      kind: import_core.ClassifierKind.CLASS
    })
  );
  return txes;
}
__name(generateMinimalModel, "generateMinimalModel");
class MockClientConnection {
  static {
    __name(this, "MockClientConnection");
  }
  handlers = [];
  _connected = true;
  hierarchy;
  model;
  transactions;
  onConnect;
  getLastHash;
  constructor(txes = []) {
    const minimalModel = txes.length === 0 ? generateMinimalModel() : txes;
    this.hierarchy = new import_core.Hierarchy();
    for (const tx of minimalModel) {
      this.hierarchy.tx(tx);
    }
    this.model = new import_core.ModelDb(this.hierarchy);
    this.transactions = new import_core.TxDb(this.hierarchy);
    for (const tx of minimalModel) {
      void this.model.tx(tx);
      void this.transactions.tx(tx);
    }
  }
  isConnected() {
    return this._connected;
  }
  setConnected(value) {
    this._connected = value;
  }
  pushHandler(handler) {
    this.handlers.push(handler);
  }
  async findAll(_class, query, options) {
    return await this.model.findAll(_class, query, options);
  }
  async searchFulltext(query, options) {
    return { docs: [] };
  }
  async tx(tx) {
    if (tx.objectSpace === import_core.default.space.Model) {
      this.hierarchy.tx(tx);
      await this.model.tx(tx);
    }
    await this.transactions.tx(tx);
    this.handlers.forEach((h) => {
      h(tx);
    });
    return {};
  }
  async loadModel(last, hash) {
    const txes = await this.transactions.findAll(import_core.default.class.Tx, {
      objectSpace: import_core.default.space.Model,
      modifiedOn: { $gt: last }
    });
    if (hash !== void 0) {
      return {
        transactions: txes,
        hash: "test-hash",
        full: false
      };
    }
    return txes;
  }
  async close() {
    this._connected = false;
  }
  async loadChunk(domain, idx) {
    return {
      idx: idx ?? 0,
      docs: [],
      finished: true
    };
  }
  async getDomainHash(domain) {
    return "test-hash";
  }
  async closeChunk(idx) {
  }
  async loadDocs(domain, docs) {
    return [];
  }
  async upload(domain, docs) {
  }
  async clean(domain, docs) {
  }
  async sendForceClose() {
  }
  async domainRequest(ctx, params, options) {
    return { domain: ctx, value: null };
  }
  simulateTransaction(tx) {
    this.handlers.forEach((h) => {
      h(tx);
    });
  }
  async simulateConnect(event = import_core.ClientConnectEvent.Connected) {
    if (this.onConnect !== void 0) {
      await this.onConnect(event, void 0, {});
    }
  }
}
async function createTestClient(initialTxes = []) {
  const connection = new MockClientConnection(initialTxes);
  const client = await (0, import_core.createClient)(async (handler) => {
    connection.pushHandler(handler);
    return connection;
  });
  return {
    client,
    connection,
    close: /* @__PURE__ */ __name(async () => {
      await client.close();
    }, "close")
  };
}
__name(createTestClient, "createTestClient");
function createTestTx(objectClass = import_core.default.class.Space, attributes = {}) {
  return {
    _id: (0, import_core.generateId)(),
    _class: import_core.default.class.TxCreateDoc,
    space: import_core.default.space.Tx,
    objectId: (0, import_core.generateId)(),
    objectClass,
    objectSpace: import_core.default.space.Model,
    modifiedBy: import_core.default.account.System,
    modifiedOn: Date.now(),
    createdOn: Date.now(),
    attributes: {
      name: "TestDoc",
      description: "",
      ...attributes
    }
  };
}
__name(createTestTx, "createTestTx");
describe("Client-Resources Integration Tests", () => {
  it("should create a full client with connection", async () => {
    const { client, connection, close } = await createTestClient();
    expect(client).toBeDefined();
    expect(client.getHierarchy()).toBeDefined();
    expect(client.getModel()).toBeDefined();
    expect(connection.isConnected()).toBe(true);
    await close();
  });
  it("should handle transactions end-to-end", async () => {
    const { client, close } = await createTestClient();
    const notifySpy = jest.fn();
    client.notify = notifySpy;
    const tx = createTestTx(import_core.default.class.Space, {
      name: "TestSpace",
      private: false,
      archived: false,
      members: []
    });
    await client.tx(tx);
    await new Promise((resolve) => setTimeout(resolve, 100));
    expect(notifySpy).toHaveBeenCalled();
    await close();
  });
  it("should handle reconnection events", async () => {
    const { connection, close } = await createTestClient();
    let connectEventReceived;
    const originalOnConnect = connection.onConnect;
    connection.onConnect = async (event, lastTx, data) => {
      connectEventReceived = event;
      await originalOnConnect?.(event, lastTx, data);
    };
    await connection.simulateConnect(import_core.ClientConnectEvent.Reconnected);
    expect(connectEventReceived).toBe(import_core.ClientConnectEvent.Reconnected);
    await close();
  });
  it("should handle model updates", async () => {
    const { client, close } = await createTestClient();
    const tx = createTestTx(import_core.default.class.Space, {
      name: "ModelSpace",
      private: false,
      archived: false,
      members: []
    });
    tx.objectSpace = import_core.default.space.Model;
    await client.tx(tx);
    const spaces = await client.findAll(import_core.default.class.Space, { name: "ModelSpace" });
    expect(spaces).toBeDefined();
    await close();
  });
  it("should handle findAll with options", async () => {
    const { client, close } = await createTestClient();
    const result = await client.findAll(import_core.default.class.Space, {}, { limit: 10, sort: { name: 1 } });
    expect(result).toBeDefined();
    expect(Array.isArray(result)).toBe(true);
    expect(result.total).toBeDefined();
    await close();
  });
  it("should handle findOne", async () => {
    const { client, close } = await createTestClient();
    const result = await client.findOne(import_core.default.class.Space, {});
    expect(result === void 0 || result !== null).toBe(true);
    await close();
  });
  it("should handle connection loss and reconnection", async () => {
    const { connection, close } = await createTestClient();
    expect(connection.isConnected()).toBe(true);
    connection.setConnected(false);
    expect(connection.isConnected()).toBe(false);
    connection.setConnected(true);
    expect(connection.isConnected()).toBe(true);
    await close();
  });
  it("should handle multiple handlers", async () => {
    const { connection, close } = await createTestClient();
    const handler1 = jest.fn();
    const handler2 = jest.fn();
    connection.pushHandler(handler1);
    connection.pushHandler(handler2);
    const tx = createTestTx();
    connection.simulateTransaction(tx);
    expect(handler1).toHaveBeenCalledWith(tx);
    expect(handler2).toHaveBeenCalledWith(tx);
    await close();
  });
  it("should handle searchFulltext", async () => {
    const { client, close } = await createTestClient();
    const query = { query: "test" };
    const options = {};
    const result = await client.searchFulltext(query, options);
    expect(result).toBeDefined();
    expect(result.docs).toBeDefined();
    expect(Array.isArray(result.docs)).toBe(true);
    await close();
  });
  it("should handle domainRequest", async () => {
    const { client, close } = await createTestClient();
    const result = await client.domainRequest("test-domain", {});
    expect(result).toBeDefined();
    await close();
  });
  it("should properly clean up on close", async () => {
    const { connection, close } = await createTestClient();
    expect(connection.isConnected()).toBe(true);
    await close();
    expect(connection.isConnected()).toBe(false);
  });
  it("should handle upgrade events", async () => {
    const { connection, close } = await createTestClient();
    let upgradeReceived = false;
    const originalOnConnect = connection.onConnect;
    connection.onConnect = async (event, lastTx, data) => {
      if (event === import_core.ClientConnectEvent.Upgraded) {
        upgradeReceived = true;
      }
      await originalOnConnect?.(event, lastTx, data);
    };
    await connection.simulateConnect(import_core.ClientConnectEvent.Upgraded);
    expect(upgradeReceived).toBe(true);
    await close();
  });
  it("should handle maintenance events", async () => {
    const { connection, close } = await createTestClient();
    let maintenanceReceived = false;
    const originalOnConnect = connection.onConnect;
    connection.onConnect = async (event, lastTx, data) => {
      if (event === import_core.ClientConnectEvent.Maintenance) {
        maintenanceReceived = true;
      }
      await originalOnConnect?.(event, lastTx, data);
    };
    await connection.simulateConnect(import_core.ClientConnectEvent.Maintenance);
    expect(maintenanceReceived).toBe(true);
    await close();
  });
  it("should handle concurrent transactions", async () => {
    const { client, connection, close } = await createTestClient();
    const notifySpy = jest.fn();
    client.notify = notifySpy;
    const tx1 = createTestTx(import_core.default.class.Space, { name: "Space1" });
    const tx2 = createTestTx(import_core.default.class.Space, { name: "Space2" });
    const tx3 = createTestTx(import_core.default.class.Space, { name: "Space3" });
    await Promise.all([client.tx(tx1), client.tx(tx2), client.tx(tx3)]);
    connection.simulateTransaction(tx1);
    connection.simulateTransaction(tx2);
    connection.simulateTransaction(tx3);
    await new Promise((resolve) => setTimeout(resolve, 100));
    expect(notifySpy.mock.calls.length).toBeGreaterThanOrEqual(3);
    await close();
  });
});
