// @hcengineering/core/lib/__tests__/collaborators.test.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var import_collaborators = require("../collaborators");
var import_memdb = require("../memdb");
var import_hierarchy = require("../hierarchy");
var import_component = __toESM(require("../component"));
describe("collaborators", () => {
  let model;
  let hierarchy;
  beforeEach(() => {
    model = new import_memdb.ModelDb(hierarchy);
    hierarchy = new import_hierarchy.Hierarchy();
  });
  describe("getClassCollaborators", () => {
    it("should return undefined when no collaborators found", () => {
      const classRef = "class:test.TestClass";
      hierarchy.getAncestors = jest.fn().mockReturnValue([classRef]);
      model.findAllSync = jest.fn().mockReturnValue([]);
      const result = (0, import_collaborators.getClassCollaborators)(model, hierarchy, classRef);
      expect(result).toBeUndefined();
    });
    it("should return collaborators for direct class", () => {
      const classRef = "class:test.TestClass";
      const collaborators = {
        _id: "collab1",
        _class: import_component.default.class.ClassCollaborators,
        space: "space1",
        modifiedOn: Date.now(),
        modifiedBy: "user1",
        attachedTo: classRef,
        attachedToClass: import_component.default.class.Class,
        collection: "collaborators"
      };
      hierarchy.getAncestors = jest.fn().mockReturnValue([classRef]);
      model.findAllSync = jest.fn().mockReturnValue([collaborators]);
      const result = (0, import_collaborators.getClassCollaborators)(model, hierarchy, classRef);
      expect(result).toBe(collaborators);
      expect(model.findAllSync).toHaveBeenCalledWith(import_component.default.class.ClassCollaborators, { attachedTo: { $in: [classRef] } });
    });
    it("should return collaborators from ancestor class", () => {
      const childClass = "class:test.ChildClass";
      const parentClass = "class:test.ParentClass";
      const grandParentClass = import_component.default.class.Doc;
      const parentCollaborators = {
        _id: "collab2",
        _class: import_component.default.class.ClassCollaborators,
        space: "space1",
        modifiedOn: Date.now(),
        modifiedBy: "user1",
        attachedTo: parentClass,
        attachedToClass: import_component.default.class.Class,
        collection: "collaborators"
      };
      hierarchy.getAncestors = jest.fn().mockReturnValue([childClass, parentClass, grandParentClass]);
      model.findAllSync = jest.fn().mockReturnValue([parentCollaborators]);
      const result = (0, import_collaborators.getClassCollaborators)(model, hierarchy, childClass);
      expect(result).toBe(parentCollaborators);
    });
    it("should return first matching ancestor collaborators", () => {
      const childClass = "class:test.ChildClass";
      const parentClass = "class:test.ParentClass";
      const grandParentClass = "class:test.GrandParentClass";
      const parentCollaborators = {
        _id: "collab3",
        _class: import_component.default.class.ClassCollaborators,
        space: "space1",
        modifiedOn: Date.now(),
        modifiedBy: "user1",
        attachedTo: parentClass,
        attachedToClass: import_component.default.class.Class,
        collection: "collaborators"
      };
      const grandParentCollaborators = {
        _id: "collab4",
        _class: import_component.default.class.ClassCollaborators,
        space: "space1",
        modifiedOn: Date.now(),
        modifiedBy: "user1",
        attachedTo: grandParentClass,
        attachedToClass: import_component.default.class.Class,
        collection: "collaborators"
      };
      hierarchy.getAncestors = jest.fn().mockReturnValue([childClass, parentClass, grandParentClass]);
      model.findAllSync = jest.fn().mockReturnValue([parentCollaborators, grandParentCollaborators]);
      const result = (0, import_collaborators.getClassCollaborators)(model, hierarchy, childClass);
      expect(result).toBe(parentCollaborators);
    });
    it("should handle single class with no ancestors", () => {
      const classRef = import_component.default.class.Doc;
      hierarchy.getAncestors = jest.fn().mockReturnValue([classRef]);
      model.findAllSync = jest.fn().mockReturnValue([]);
      const result = (0, import_collaborators.getClassCollaborators)(model, hierarchy, classRef);
      expect(result).toBeUndefined();
    });
    it("should handle empty ancestors list", () => {
      const classRef = "class:test.TestClass";
      hierarchy.getAncestors = jest.fn().mockReturnValue([]);
      model.findAllSync = jest.fn().mockReturnValue([]);
      const result = (0, import_collaborators.getClassCollaborators)(model, hierarchy, classRef);
      expect(result).toBeUndefined();
    });
    it("should properly query model with $in operator", () => {
      const childClass = "class:test.ChildClass";
      const parentClass = "class:test.ParentClass";
      const ancestors = [childClass, parentClass, import_component.default.class.Doc];
      hierarchy.getAncestors = jest.fn().mockReturnValue(ancestors);
      model.findAllSync = jest.fn().mockReturnValue([]);
      (0, import_collaborators.getClassCollaborators)(model, hierarchy, childClass);
      expect(model.findAllSync).toHaveBeenCalledWith(import_component.default.class.ClassCollaborators, { attachedTo: { $in: ancestors } });
    });
    it("should iterate through ancestors in order", () => {
      const class1 = "class:test.Class1";
      const class2 = "class:test.Class2";
      const class3 = "class:test.Class3";
      const collab2 = {
        _id: "collab5",
        _class: import_component.default.class.ClassCollaborators,
        space: "space1",
        modifiedOn: Date.now(),
        modifiedBy: "user1",
        attachedTo: class2,
        attachedToClass: import_component.default.class.Class,
        collection: "collaborators"
      };
      const collab3 = {
        _id: "collab6",
        _class: import_component.default.class.ClassCollaborators,
        space: "space1",
        modifiedOn: Date.now(),
        modifiedBy: "user1",
        attachedTo: class3,
        attachedToClass: import_component.default.class.Class,
        collection: "collaborators"
      };
      hierarchy.getAncestors = jest.fn().mockReturnValue([class1, class2, class3]);
      model.findAllSync = jest.fn().mockReturnValue([collab2, collab3]);
      const result = (0, import_collaborators.getClassCollaborators)(model, hierarchy, class1);
      expect(result).toBe(collab2);
    });
  });
});
