// @hcengineering/core/lib/__tests__/hierarchy.test.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var import_classes = require("../classes");
var import_tx = require("../tx");
var import_component = __toESM(require("../component"));
var import_hierarchy = require("../hierarchy");
var Proxy2 = __toESM(require("../proxy"));
var import_minmodel = require("./minmodel");
const txes = (0, import_minmodel.genMinModel)();
function prepare() {
  const hierarchy = new import_hierarchy.Hierarchy();
  for (const tx of txes) hierarchy.tx(tx);
  return hierarchy;
}
__name(prepare, "prepare");
describe("hierarchy", () => {
  it("should build hierarchy", async () => {
    const hierarchy = prepare();
    const ancestors = hierarchy.getAncestors(import_component.default.class.TxCreateDoc);
    expect(ancestors).toContain(import_component.default.class.Tx);
  });
  it("isDerived", async () => {
    const hierarchy = prepare();
    const derived = hierarchy.isDerived(import_component.default.class.Space, import_component.default.class.Doc);
    expect(derived).toBeTruthy();
    const notDerived = hierarchy.isDerived(import_component.default.class.Space, import_component.default.class.Class);
    expect(notDerived).not.toBeTruthy();
  });
  it("isImplements", async () => {
    const hierarchy = prepare();
    let isImplements = hierarchy.isImplements(import_minmodel.test.class.Task, import_minmodel.test.interface.WithState);
    expect(isImplements).toBeTruthy();
    isImplements = hierarchy.isImplements(import_minmodel.test.class.TaskCheckItem, import_minmodel.test.interface.WithState);
    expect(isImplements).toBeTruthy();
    const notImplements = hierarchy.isImplements(import_component.default.class.Space, import_minmodel.test.interface.WithState);
    expect(notImplements).not.toBeTruthy();
  });
  it("getClass", async () => {
    const hierarchy = prepare();
    const data = hierarchy.getClass(import_component.default.class.TxCreateDoc);
    expect(data).toMatchObject(txes.find((p) => p.objectId === import_component.default.class.TxCreateDoc).attributes);
    const notExistClass = "class:test.MyClass";
    expect(() => hierarchy.getClass(notExistClass)).toThrowError("class not found: " + notExistClass);
  });
  it("getDomain", async () => {
    const hierarchy = prepare();
    const txDomain = hierarchy.getDomain(import_component.default.class.TxCreateDoc);
    expect(txDomain).toBe("tx");
    const modelDomain = hierarchy.getDomain(import_component.default.class.Class);
    expect(modelDomain).toBe("model");
  });
  it("should create Mixin proxy", async () => {
    const spyProxy = jest.spyOn(Proxy2, "_createMixinProxy");
    const hierarchy = prepare();
    hierarchy.as(txes[0], import_minmodel.test.mixin.TestMixin);
    expect(spyProxy).toBeCalledTimes(1);
    hierarchy.as(txes[0], import_minmodel.test.mixin.TestMixin);
    expect(spyProxy).toBeCalledTimes(1);
    spyProxy.mockReset();
    spyProxy.mockRestore();
  });
  it("should call static methods", async () => {
    const spyToDoc = jest.spyOn(Proxy2, "_toDoc");
    import_hierarchy.Hierarchy.toDoc(txes[0]);
    expect(spyToDoc).toBeCalledTimes(1);
    spyToDoc.mockReset();
    spyToDoc.mockRestore();
    const spyMixinClass = jest.spyOn(Proxy2, "_mixinClass");
    import_hierarchy.Hierarchy.mixinClass(txes[0]);
    expect(spyMixinClass).toBeCalledTimes(1);
    spyMixinClass.mockImplementationOnce(() => void 0).mockImplementationOnce(() => import_minmodel.test.mixin.TestMixin);
    let result = import_hierarchy.Hierarchy.mixinOrClass(txes[0]);
    expect(result).toStrictEqual(txes[0]._class);
    result = import_hierarchy.Hierarchy.mixinOrClass(txes[0]);
    expect(result).toStrictEqual(import_minmodel.test.mixin.TestMixin);
    expect(spyMixinClass).toBeCalledTimes(3);
    spyMixinClass.mockReset();
    spyMixinClass.mockRestore();
  });
  it("getAncestors should return array directly", async () => {
    const hierarchy = prepare();
    const ancestors = hierarchy.getAncestors(import_component.default.class.TxCreateDoc);
    expect(Array.isArray(ancestors)).toBeTruthy();
    expect(ancestors).toContain(import_component.default.class.TxCreateDoc);
    expect(ancestors).toContain(import_component.default.class.TxCUD);
    expect(ancestors).toContain(import_component.default.class.Tx);
    expect(ancestors).toContain(import_component.default.class.Doc);
    expect(ancestors).toContain(import_component.default.class.Obj);
    const indexTx = ancestors.indexOf(import_component.default.class.Tx);
    const indexDoc = ancestors.indexOf(import_component.default.class.Doc);
    expect(indexDoc).toBeGreaterThan(indexTx);
  });
  it("isDerived should work with array-based ancestors", async () => {
    const hierarchy = prepare();
    expect(hierarchy.isDerived(import_component.default.class.TxCreateDoc, import_component.default.class.Tx)).toBeTruthy();
    expect(hierarchy.isDerived(import_component.default.class.TxCreateDoc, import_component.default.class.TxCUD)).toBeTruthy();
    expect(hierarchy.isDerived(import_component.default.class.TxCreateDoc, import_component.default.class.Doc)).toBeTruthy();
    expect(hierarchy.isDerived(import_component.default.class.TxCreateDoc, import_component.default.class.Obj)).toBeTruthy();
    expect(hierarchy.isDerived(import_component.default.class.TxCreateDoc, import_component.default.class.TxCreateDoc)).toBeTruthy();
    expect(hierarchy.isDerived(import_component.default.class.TxCreateDoc, import_component.default.class.Space)).toBeFalsy();
    expect(hierarchy.isDerived(import_component.default.class.Space, import_component.default.class.Tx)).toBeFalsy();
    expect(hierarchy.isDerived(import_minmodel.test.class.TestComment, import_component.default.class.AttachedDoc)).toBeTruthy();
    expect(hierarchy.isDerived(import_minmodel.test.mixin.TaskMixinTodos, import_minmodel.test.class.Task)).toBeTruthy();
  });
  it("should handle deep inheritance chains efficiently", async () => {
    const hierarchy = prepare();
    const ancestors = hierarchy.getAncestors(import_component.default.class.TxCreateDoc);
    expect(ancestors.length).toBeGreaterThanOrEqual(5);
    expect(hierarchy.isDerived(import_component.default.class.TxCreateDoc, import_component.default.class.TxCUD)).toBeTruthy();
    expect(hierarchy.isDerived(import_component.default.class.TxCUD, import_component.default.class.Tx)).toBeTruthy();
    expect(hierarchy.isDerived(import_component.default.class.Tx, import_component.default.class.Doc)).toBeTruthy();
    expect(hierarchy.isDerived(import_component.default.class.Doc, import_component.default.class.Obj)).toBeTruthy();
  });
  it("getClassifierProp and setClassifierProp should work with Map", async () => {
    const hierarchy = prepare();
    hierarchy.setClassifierProp(import_component.default.class.Space, "testProp", "testValue");
    const value = hierarchy.getClassifierProp(import_component.default.class.Space, "testProp");
    expect(value).toBe("testValue");
    hierarchy.setClassifierProp(import_component.default.class.Space, "testProp", "updatedValue");
    const updatedValue = hierarchy.getClassifierProp(import_component.default.class.Space, "testProp");
    expect(updatedValue).toBe("updatedValue");
  });
  it("should handle multiple properties per classifier", async () => {
    const hierarchy = prepare();
    hierarchy.setClassifierProp(import_component.default.class.Space, "prop1", "value1");
    hierarchy.setClassifierProp(import_component.default.class.Space, "prop2", "value2");
    hierarchy.setClassifierProp(import_component.default.class.Space, "prop3", 42);
    hierarchy.setClassifierProp(import_component.default.class.Space, "prop4", { nested: "object" });
    expect(hierarchy.getClassifierProp(import_component.default.class.Space, "prop1")).toBe("value1");
    expect(hierarchy.getClassifierProp(import_component.default.class.Space, "prop2")).toBe("value2");
    expect(hierarchy.getClassifierProp(import_component.default.class.Space, "prop3")).toBe(42);
    expect(hierarchy.getClassifierProp(import_component.default.class.Space, "prop4")).toEqual({ nested: "object" });
    expect(hierarchy.getClassifierProp(import_component.default.class.Space, "nonExistent")).toBeUndefined();
  });
  it("should isolate properties between different classifiers", async () => {
    const hierarchy = prepare();
    hierarchy.setClassifierProp(import_component.default.class.Space, "name", "Space");
    hierarchy.setClassifierProp(import_component.default.class.Doc, "name", "Doc");
    hierarchy.setClassifierProp(import_minmodel.test.class.Task, "name", "Task");
    expect(hierarchy.getClassifierProp(import_component.default.class.Space, "name")).toBe("Space");
    expect(hierarchy.getClassifierProp(import_component.default.class.Doc, "name")).toBe("Doc");
    expect(hierarchy.getClassifierProp(import_minmodel.test.class.Task, "name")).toBe("Task");
  });
  it("should handle property updates without creating new objects", async () => {
    const hierarchy = prepare();
    hierarchy.setClassifierProp(import_component.default.class.Space, "counter", 0);
    for (let i = 1; i <= 100; i++) {
      hierarchy.setClassifierProp(import_component.default.class.Space, "counter", i);
    }
    expect(hierarchy.getClassifierProp(import_component.default.class.Space, "counter")).toBe(100);
  });
  it("should handle interface implementation checks correctly", async () => {
    const hierarchy = prepare();
    expect(hierarchy.isImplements(import_minmodel.test.class.Task, import_minmodel.test.interface.WithState)).toBeTruthy();
    expect(hierarchy.isImplements(import_minmodel.test.class.Task, import_minmodel.test.interface.DummyWithState)).toBeTruthy();
    expect(hierarchy.isImplements(import_minmodel.test.class.TaskCheckItem, import_minmodel.test.interface.WithState)).toBeTruthy();
    expect(hierarchy.isImplements(import_component.default.class.Space, import_minmodel.test.interface.WithState)).toBeFalsy();
  });
  it("should maintain consistency after multiple hierarchy operations", async () => {
    const hierarchy = prepare();
    const ancestors1 = hierarchy.getAncestors(import_minmodel.test.class.Task);
    const isDerived1 = hierarchy.isDerived(import_minmodel.test.class.Task, import_component.default.class.Doc);
    hierarchy.setClassifierProp(import_minmodel.test.class.Task, "test", "value");
    const ancestors2 = hierarchy.getAncestors(import_minmodel.test.class.Task);
    const isDerived2 = hierarchy.isDerived(import_minmodel.test.class.Task, import_component.default.class.Doc);
    expect(ancestors1).toEqual(ancestors2);
    expect(isDerived1).toBe(isDerived2);
    expect(isDerived2).toBeTruthy();
  });
  it("should handle getDescendants correctly", async () => {
    const hierarchy = prepare();
    const descendants = hierarchy.getDescendants(import_component.default.class.Doc);
    expect(descendants).toContain(import_component.default.class.Space);
    expect(descendants).toContain(import_component.default.class.Tx);
    expect(descendants).toContain(import_minmodel.test.class.Task);
    expect(Array.isArray(descendants)).toBeTruthy();
  });
  it("should work with getBaseClass", async () => {
    const hierarchy = prepare();
    const baseClass = hierarchy.getBaseClass(import_minmodel.test.mixin.TaskMixinTodos);
    expect(baseClass).toBe(import_minmodel.test.class.Task);
    const baseClass2 = hierarchy.getBaseClass(import_minmodel.test.class.Task);
    expect(baseClass2).toBe(import_minmodel.test.class.Task);
  });
  it("should handle getAllAttributes correctly", async () => {
    const hierarchy = prepare();
    const attributes = hierarchy.getAllAttributes(import_component.default.class.TxCreateDoc);
    expect(attributes instanceof Map).toBeTruthy();
    const attributesTo = hierarchy.getAllAttributes(import_component.default.class.TxCreateDoc, import_component.default.class.Tx);
    expect(attributesTo instanceof Map).toBeTruthy();
  });
  it("should maintain immutability of returned ancestors array", async () => {
    const hierarchy = prepare();
    const ancestors = hierarchy.getAncestors(import_minmodel.test.class.Task);
    const originalLength = ancestors.length;
    expect(ancestors.length).toBe(originalLength);
    expect(ancestors).toContain(import_component.default.class.Doc);
  });
  it("should handle performance for multiple isDerived checks", async () => {
    const hierarchy = prepare();
    const startTime = Date.now();
    for (let i = 0; i < 1e3; i++) {
      hierarchy.isDerived(import_component.default.class.TxCreateDoc, import_component.default.class.Tx);
      hierarchy.isDerived(import_minmodel.test.class.Task, import_component.default.class.Doc);
      hierarchy.isDerived(import_minmodel.test.class.TaskCheckItem, import_component.default.class.AttachedDoc);
    }
    const endTime = Date.now();
    expect(endTime - startTime).toBeLessThan(100);
  });
  it("should handle findClass and hasClass correctly", async () => {
    const hierarchy = prepare();
    const foundClass = hierarchy.findClass(import_component.default.class.Space);
    expect(foundClass).toBeDefined();
    expect(foundClass?._id).toBe(import_component.default.class.Space);
    const notFoundClass = hierarchy.findClass("class:NonExistent");
    expect(notFoundClass).toBeUndefined();
    expect(hierarchy.hasClass(import_component.default.class.Space)).toBeTruthy();
    expect(hierarchy.hasClass("class:NonExistent")).toBeFalsy();
    expect(hierarchy.hasClass(import_minmodel.test.interface.WithState)).toBeFalsy();
  });
  it("should handle getClassOrInterface correctly", async () => {
    const hierarchy = prepare();
    const spaceClass = hierarchy.getClassOrInterface(import_component.default.class.Space);
    expect(spaceClass._id).toBe(import_component.default.class.Space);
    const withStateInterface = hierarchy.getClassOrInterface(import_minmodel.test.interface.WithState);
    expect(withStateInterface._id).toBe(import_minmodel.test.interface.WithState);
    expect(() => hierarchy.getClassOrInterface("class:NonExistent")).toThrowError(
      "class not found: class:NonExistent"
    );
  });
  it("should handle getInterface correctly", async () => {
    const hierarchy = prepare();
    const withStateInterface = hierarchy.getInterface(import_minmodel.test.interface.WithState);
    expect(withStateInterface._id).toBe(import_minmodel.test.interface.WithState);
    expect(() => hierarchy.getInterface("interface:NonExistent")).toThrowError(
      "interface not found: interface:NonExistent"
    );
    expect(() => hierarchy.getInterface(import_component.default.class.Space)).toThrowError();
  });
  it("should handle isMixin correctly", async () => {
    const hierarchy = prepare();
    expect(hierarchy.isMixin(import_minmodel.test.mixin.TestMixin)).toBeTruthy();
    expect(hierarchy.isMixin(import_minmodel.test.mixin.TaskMixinTodos)).toBeTruthy();
    expect(hierarchy.isMixin(import_component.default.class.AttachedDoc)).toBeTruthy();
    expect(hierarchy.isMixin(import_component.default.class.Space)).toBeFalsy();
    expect(hierarchy.isMixin(import_minmodel.test.class.Task)).toBeFalsy();
    expect(hierarchy.isMixin("mixin:NonExistent")).toBeFalsy();
  });
  it("should handle as and asIf with mixins", async () => {
    const hierarchy = prepare();
    const doc = {
      _id: "doc1",
      _class: import_component.default.class.Doc,
      space: "space1",
      modifiedOn: 0,
      modifiedBy: "user1"
    };
    const withMixin = hierarchy.as(doc, import_minmodel.test.mixin.TestMixin);
    expect(withMixin).toBeDefined();
    expect(withMixin._id).toBe(doc._id);
    const asIfResult = hierarchy.asIf(doc, import_minmodel.test.mixin.TestMixin);
    expect(asIfResult).toBeUndefined();
    expect(hierarchy.asIf(void 0, import_minmodel.test.mixin.TestMixin)).toBeUndefined();
  });
  it("should handle asIfArray correctly", async () => {
    const hierarchy = prepare();
    const docs = [
      { _id: "doc1", _class: import_component.default.class.Doc, space: "space1", modifiedOn: 0, modifiedBy: "user1" },
      { _id: "doc2", _class: import_component.default.class.Doc, space: "space1", modifiedOn: 0, modifiedBy: "user1" }
    ];
    const result = hierarchy.asIfArray(docs, import_minmodel.test.mixin.TestMixin);
    expect(Array.isArray(result)).toBeTruthy();
    expect(result.length).toBe(0);
  });
  it("should handle classHierarchyMixin correctly", async () => {
    const hierarchy = prepare();
    const result = hierarchy.classHierarchyMixin(import_minmodel.test.mixin.TaskMixinTodos, import_minmodel.test.mixin.TestMixin);
    expect(result === void 0 || typeof result === "object").toBeTruthy();
    const filtered = hierarchy.classHierarchyMixin(import_minmodel.test.class.Task, import_minmodel.test.mixin.TestMixin, (m) => m !== void 0);
    expect(filtered === void 0 || typeof filtered === "object").toBeTruthy();
  });
  it("should handle findClassOrMixinMixin correctly", async () => {
    const hierarchy = prepare();
    const doc = {
      _id: "doc1",
      _class: import_minmodel.test.class.Task,
      space: "space1",
      modifiedOn: 0,
      modifiedBy: "user1"
    };
    const result = hierarchy.findClassOrMixinMixin(doc, import_minmodel.test.mixin.TestMixin);
    expect(result === void 0 || typeof result === "object").toBeTruthy();
  });
  it("should handle findMixinMixins correctly", async () => {
    const hierarchy = prepare();
    const doc = {
      _id: "doc1",
      _class: import_minmodel.test.class.Task,
      space: "space1",
      modifiedOn: 0,
      modifiedBy: "user1"
    };
    const results = hierarchy.findMixinMixins(doc, import_minmodel.test.mixin.TestMixin);
    expect(Array.isArray(results)).toBeTruthy();
  });
  it("should handle findAllMixins correctly", async () => {
    const hierarchy = prepare();
    const doc = {
      _id: "doc1",
      _class: import_minmodel.test.class.Task,
      space: "space1",
      modifiedOn: 0,
      modifiedBy: "user1"
    };
    const mixins = hierarchy.findAllMixins(doc);
    expect(Array.isArray(mixins)).toBeTruthy();
  });
  it("should handle findDomain correctly", async () => {
    const hierarchy = prepare();
    const domain = hierarchy.findDomain(import_component.default.class.Space);
    expect(domain).toBe("model");
    const noDomain = hierarchy.findDomain("class:NonExistent");
    expect(noDomain).toBeUndefined();
  });
  it("should handle getParentClass correctly", async () => {
    const hierarchy = prepare();
    const parent = hierarchy.getParentClass(import_component.default.class.Space);
    expect(parent).toBeDefined();
    const parentDomain = hierarchy.findDomain(parent);
    expect(parentDomain === "model" || parent === import_component.default.class.Space).toBeTruthy();
  });
  it("should handle getAttribute correctly", async () => {
    const hierarchy = prepare();
    expect(() => hierarchy.getAttribute(import_component.default.class.Space, "nonExistentAttr")).toThrowError(
      "attribute not found: nonExistentAttr"
    );
  });
  it("should handle findAttribute correctly", async () => {
    const hierarchy = prepare();
    const attr = hierarchy.findAttribute(import_component.default.class.Space, "nonExistentAttr");
    expect(attr).toBeUndefined();
  });
  it("should handle getOwnAttributes correctly", async () => {
    const hierarchy = prepare();
    const attrs = hierarchy.getOwnAttributes(import_component.default.class.Space);
    expect(attrs instanceof Map).toBeTruthy();
  });
  it("should handle updateLookupMixin correctly", async () => {
    const hierarchy = prepare();
    const doc = {
      _id: "doc1",
      _class: import_minmodel.test.class.Task,
      space: "space1",
      modifiedOn: 0,
      modifiedBy: "user1"
    };
    const result1 = hierarchy.updateLookupMixin(import_minmodel.test.class.Task, doc);
    expect(result1).toBeDefined();
    const docWithLookup = {
      ...doc,
      $lookup: {}
    };
    const result2 = hierarchy.updateLookupMixin(import_minmodel.test.class.Task, docWithLookup);
    expect(result2).toBeDefined();
  });
  it("should handle clone correctly", async () => {
    const hierarchy = prepare();
    const obj = {
      _id: "doc1",
      _class: import_component.default.class.Doc,
      space: "space1",
      modifiedOn: 0,
      modifiedBy: "user1",
      nested: {
        value: "test"
      }
    };
    const cloned = hierarchy.clone(obj);
    expect(cloned).toEqual(obj);
    expect(cloned).not.toBe(obj);
    expect(cloned.nested).not.toBe(obj.nested);
  });
  it("should handle domains correctly", async () => {
    const hierarchy = prepare();
    const domains = hierarchy.domains();
    expect(Array.isArray(domains)).toBeTruthy();
    expect(domains).toContain("model");
    expect(domains).toContain("tx");
    const uniqueDomains = [...new Set(domains)];
    expect(domains.length).toBe(uniqueDomains.length);
  });
  it("should handle getAncestors error case", async () => {
    const hierarchy = prepare();
    expect(() => hierarchy.getAncestors("class:NonExistent")).toThrowError(
      "ancestors not found: class:NonExistent"
    );
  });
  it("should handle getDescendants error case", async () => {
    const hierarchy = prepare();
    expect(() => hierarchy.getDescendants("class:NonExistent")).toThrowError(
      "descendants not found: class:NonExistent"
    );
  });
  it("should handle getDomain error case", async () => {
    const hierarchy = prepare();
    expect(() => hierarchy.getDomain("class:NonExistent")).toThrowError("domain not found: class:NonExistent");
  });
  it("should handle static hasMixin correctly", async () => {
    const doc = {
      _id: "doc1",
      _class: import_minmodel.test.class.Task,
      space: "space1",
      modifiedOn: 0,
      modifiedBy: "user1",
      [import_minmodel.test.mixin.TestMixin]: { arr: [] }
    };
    expect(import_hierarchy.Hierarchy.hasMixin(doc, import_minmodel.test.mixin.TestMixin)).toBeTruthy();
    expect(import_hierarchy.Hierarchy.hasMixin(doc, "mixin:Other")).toBeFalsy();
  });
  it("should maintain descendants list correctly", async () => {
    const hierarchy = prepare();
    const docDescendants = hierarchy.getDescendants(import_component.default.class.Doc);
    expect(docDescendants.length).toBeGreaterThan(5);
    const objDescendants = hierarchy.getDescendants(import_component.default.class.Obj);
    expect(objDescendants.length).toBeGreaterThanOrEqual(docDescendants.length);
    const txCreateDescendants = hierarchy.getDescendants(import_component.default.class.TxCreateDoc);
    expect(Array.isArray(txCreateDescendants)).toBeTruthy();
  });
  it("should handle interface extends chains", async () => {
    const hierarchy = prepare();
    const dummyAncestors = hierarchy.getAncestors(import_minmodel.test.interface.DummyWithState);
    expect(dummyAncestors).toContain(import_minmodel.test.interface.WithState);
  });
  it("should handle complex mixin hierarchies", async () => {
    const hierarchy = prepare();
    expect(hierarchy.isDerived(import_minmodel.test.mixin.TaskMixinTodos, import_minmodel.test.class.Task)).toBeTruthy();
    expect(hierarchy.isDerived(import_minmodel.test.mixin.TaskMixinTodos, import_component.default.class.Doc)).toBeTruthy();
    expect(hierarchy.isMixin(import_minmodel.test.mixin.TaskMixinTodos)).toBeTruthy();
    const baseClass = hierarchy.getBaseClass(import_minmodel.test.mixin.TaskMixinTodos);
    expect(baseClass).toBe(import_minmodel.test.class.Task);
  });
  it("should handle getAllAttributes with traverse callback", async () => {
    const hierarchy = prepare();
    const traversed = [];
    const attributes = hierarchy.getAllAttributes(import_component.default.class.TxCreateDoc, void 0, (name, attr) => {
      traversed.push({ name, attrId: attr._id });
    });
    expect(attributes instanceof Map).toBeTruthy();
    if (attributes.size > 0) {
      expect(traversed.length).toBeGreaterThan(0);
    }
  });
  it("should handle property deletions implicitly through cache invalidation", async () => {
    const hierarchy = prepare();
    hierarchy.setClassifierProp(import_component.default.class.Space, "cached", "value");
    expect(hierarchy.getClassifierProp(import_component.default.class.Space, "cached")).toBe("value");
    expect(hierarchy.getClassifierProp(import_component.default.class.Space, "cached")).toBe("value");
  });
  it("should verify memory optimization: no duplicate ancestor storage", async () => {
    const hierarchy = prepare();
    const ancestors1 = hierarchy.getAncestors(import_component.default.class.TxCreateDoc);
    const ancestors2 = hierarchy.getAncestors(import_component.default.class.TxUpdateDoc);
    const ancestors3 = hierarchy.getAncestors(import_component.default.class.TxRemoveDoc);
    const commonAncestors = ancestors1.filter((a) => ancestors2.includes(a) && ancestors3.includes(a));
    expect(commonAncestors).toContain(import_component.default.class.TxCUD);
    expect(commonAncestors).toContain(import_component.default.class.Tx);
    expect(commonAncestors).toContain(import_component.default.class.Doc);
    expect(ancestors1).not.toBe(ancestors2);
    expect(ancestors2).not.toBe(ancestors3);
  });
  it("should handle classHierarchyMixin with filter that returns false", async () => {
    const hierarchy = prepare();
    const clazz = hierarchy.getClass(import_component.default.class.Doc);
    const withMixin = hierarchy.as(clazz, import_minmodel.test.mixin.TestMixin);
    expect(withMixin).toBeDefined();
    const result = hierarchy.classHierarchyMixin(
      import_component.default.class.Doc,
      import_minmodel.test.mixin.TestMixin,
      (m) => false
      // Filter that always returns false
    );
    expect(result).toBeUndefined();
  });
  it("should handle findClassOrMixinMixin with document having mixins", async () => {
    const hierarchy = prepare();
    const doc = {
      _id: "doc1",
      _class: import_minmodel.test.class.Task,
      space: "space1",
      modifiedOn: 0,
      modifiedBy: "user1",
      [import_minmodel.test.mixin.TestMixin]: { arr: ["item1"] }
    };
    const result = hierarchy.findClassOrMixinMixin(doc, import_minmodel.test.mixin.TestMixin);
    expect(result === void 0 || typeof result === "object").toBeTruthy();
  });
  it("should handle findMixinMixins with document having multiple mixins", async () => {
    const hierarchy = prepare();
    const taskClass = hierarchy.getClass(import_minmodel.test.class.Task);
    hierarchy.as(taskClass, import_minmodel.test.mixin.TestMixin);
    const doc = {
      _id: "doc1",
      _class: import_minmodel.test.class.Task,
      space: "space1",
      modifiedOn: 0,
      modifiedBy: "user1",
      [import_minmodel.test.mixin.TestMixin]: { arr: ["item1"] }
    };
    const results = hierarchy.findMixinMixins(doc, import_minmodel.test.mixin.TestMixin);
    expect(Array.isArray(results)).toBeTruthy();
  });
  it("should handle findAllMixins with document having mixins", async () => {
    const hierarchy = prepare();
    const doc = {
      _id: "doc1",
      _class: import_minmodel.test.class.Task,
      space: "space1",
      modifiedOn: 0,
      modifiedBy: "user1",
      [import_minmodel.test.mixin.TestMixin]: { arr: ["item1"] }
    };
    const mixins = hierarchy.findAllMixins(doc);
    expect(Array.isArray(mixins)).toBeTruthy();
    if (mixins.length > 0) {
      expect(mixins).toContain(import_minmodel.test.mixin.TestMixin);
    }
  });
  it("should handle tx with TxMixin", async () => {
    const hierarchy = new import_hierarchy.Hierarchy();
    for (const tx of txes) {
      hierarchy.tx(tx);
    }
    const txFactory = new import_tx.TxFactory(import_component.default.account.System);
    const mixinTx = txFactory.createTxMixin(
      import_component.default.class.Space,
      import_component.default.class.Class,
      import_component.default.space.Model,
      import_minmodel.test.mixin.TestMixin,
      { arr: ["test"] }
    );
    hierarchy.tx(mixinTx);
    const spaceClass = hierarchy.getClass(import_component.default.class.Space);
    expect(spaceClass).toBeDefined();
  });
  it("should handle txUpdateDoc with Attribute", async () => {
    const hierarchy = new import_hierarchy.Hierarchy();
    for (const tx of txes) {
      hierarchy.tx(tx);
    }
    const txFactory = new import_tx.TxFactory(import_component.default.account.System);
    const attrId = "attr:test";
    const createAttrTx = txFactory.createTxCreateDoc(
      import_component.default.class.Attribute,
      import_component.default.space.Model,
      {
        attributeOf: import_component.default.class.Space,
        name: "testAttr",
        type: { _class: "class:core.Type" }
      },
      attrId
    );
    hierarchy.tx(createAttrTx);
    const updateAttrTx = txFactory.createTxUpdateDoc(import_component.default.class.Attribute, import_component.default.space.Model, attrId, {
      name: "updatedAttr"
    });
    hierarchy.tx(updateAttrTx);
    const attr = hierarchy.findAttribute(import_component.default.class.Space, "updatedAttr");
    expect(attr).toBeDefined();
  });
  it("should handle txRemoveDoc with Attribute", async () => {
    const hierarchy = new import_hierarchy.Hierarchy();
    for (const tx of txes) {
      hierarchy.tx(tx);
    }
    const txFactory = new import_tx.TxFactory(import_component.default.account.System);
    const attrId = "attr:test";
    const createAttrTx = txFactory.createTxCreateDoc(
      import_component.default.class.Attribute,
      import_component.default.space.Model,
      {
        attributeOf: import_component.default.class.Space,
        name: "testAttr",
        type: { _class: "class:core.Type" }
      },
      attrId
    );
    hierarchy.tx(createAttrTx);
    let attr = hierarchy.findAttribute(import_component.default.class.Space, "testAttr");
    expect(attr).toBeDefined();
    const removeAttrTx = txFactory.createTxRemoveDoc(import_component.default.class.Attribute, import_component.default.space.Model, attrId);
    hierarchy.tx(removeAttrTx);
    attr = hierarchy.findAttribute(import_component.default.class.Space, "testAttr");
    expect(attr).toBeUndefined();
  });
  it("should handle txUpdateDoc with Classifier", async () => {
    const hierarchy = new import_hierarchy.Hierarchy();
    for (const tx of txes) {
      hierarchy.tx(tx);
    }
    const txFactory = new import_tx.TxFactory(import_component.default.account.System);
    const updateClassTx = txFactory.createTxUpdateDoc(import_component.default.class.Class, import_component.default.space.Model, import_component.default.class.Space, {
      label: "Updated Space"
    });
    hierarchy.tx(updateClassTx);
    const spaceClass = hierarchy.getClass(import_component.default.class.Space);
    expect(spaceClass.label).toBe("Updated Space");
  });
  it("should handle txRemoveDoc with Classifier", async () => {
    const hierarchy = new import_hierarchy.Hierarchy();
    for (const tx of txes) {
      hierarchy.tx(tx);
    }
    const txFactory = new import_tx.TxFactory(import_component.default.account.System);
    const newClassId = "class:test.NewClass";
    const createClassTx = txFactory.createTxCreateDoc(
      import_component.default.class.Class,
      import_component.default.space.Model,
      {
        label: "NewClass",
        extends: import_component.default.class.Doc,
        kind: import_classes.ClassifierKind.CLASS,
        domain: import_classes.DOMAIN_MODEL
      },
      newClassId
    );
    hierarchy.tx(createClassTx);
    expect(hierarchy.hasClass(newClassId)).toBeTruthy();
    const removeClassTx = txFactory.createTxRemoveDoc(import_component.default.class.Class, import_component.default.space.Model, newClassId);
    hierarchy.tx(removeClassTx);
    expect(hierarchy.hasClass(newClassId)).toBeFalsy();
  });
  it("should handle updateLookupMixin with lookup containing mixins", async () => {
    const hierarchy = prepare();
    const doc = {
      _id: "doc1",
      _class: import_minmodel.test.class.Task,
      space: "space1",
      modifiedOn: 0,
      modifiedBy: "user1",
      $lookup: {
        space: {
          _id: "space1",
          _class: import_component.default.class.Space,
          name: "Test Space",
          description: "",
          private: false,
          members: [],
          archived: false,
          modifiedOn: 0,
          modifiedBy: "user1"
        }
      }
    };
    const options = {
      lookup: {
        space: import_component.default.class.Space
      }
    };
    const result = hierarchy.updateLookupMixin(import_minmodel.test.class.Task, doc, options);
    expect(result).toBeDefined();
    expect(result.$lookup).toBeDefined();
  });
  it("should handle updateLookupMixin with _id lookup containing mixins", async () => {
    const hierarchy = prepare();
    const doc = {
      _id: "doc1",
      _class: import_minmodel.test.class.Task,
      space: "space1",
      modifiedOn: 0,
      modifiedBy: "user1",
      $lookup: {
        _id: {
          _class: import_component.default.class.Space,
          _id: "space1",
          name: "Test",
          description: "",
          private: false,
          members: [],
          archived: false,
          modifiedOn: 0,
          modifiedBy: "user1"
        }
      }
    };
    const options = {
      lookup: {
        _id: {
          _class: import_minmodel.test.mixin.TestMixin
        }
      }
    };
    const result = hierarchy.updateLookupMixin(import_minmodel.test.class.Task, doc, options);
    expect(result).toBeDefined();
  });
  it("should handle updateLookupMixin with array in _id lookup", async () => {
    const hierarchy = prepare();
    const doc = {
      _id: "doc1",
      _class: import_minmodel.test.class.Task,
      space: "space1",
      modifiedOn: 0,
      modifiedBy: "user1",
      $lookup: {
        _id: [
          {
            _class: import_component.default.class.Space,
            _id: "space1",
            name: "Test",
            description: "",
            private: false,
            members: [],
            archived: false,
            modifiedOn: 0,
            modifiedBy: "user1"
          }
        ]
      }
    };
    const options = {
      lookup: {
        _id: {
          _class: [import_minmodel.test.mixin.TestMixin]
        }
      }
    };
    const result = hierarchy.updateLookupMixin(import_minmodel.test.class.Task, doc, options);
    expect(result).toBeDefined();
  });
  it("should handle updateLookupMixin with null lookup value", async () => {
    const hierarchy = prepare();
    const doc = {
      _id: "doc1",
      _class: import_minmodel.test.class.Task,
      space: "space1",
      modifiedOn: 0,
      modifiedBy: "user1",
      $lookup: {
        space: null
      }
    };
    const options = {
      lookup: {
        space: import_minmodel.test.mixin.TestMixin
      }
    };
    const result = hierarchy.updateLookupMixin(import_minmodel.test.class.Task, doc, options);
    expect(result).toBeDefined();
    expect(result.$lookup?.space).toBeNull();
  });
  it("should handle tx with non-matching class", async () => {
    const hierarchy = new import_hierarchy.Hierarchy();
    for (const tx of txes) {
      hierarchy.tx(tx);
    }
    const txFactory = new import_tx.TxFactory(import_component.default.account.System);
    const spaceTx = txFactory.createTxCreateDoc(import_component.default.class.Space, import_component.default.space.Model, {
      name: "Test Space",
      description: "",
      private: false,
      members: [],
      archived: false
    });
    hierarchy.tx(spaceTx);
    expect(true).toBeTruthy();
  });
});
